{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fc5c48b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from data_preprocessing import  separate_tf_genes, prepare_for_inference\n",
    "from tf_utils import map_tf_ids\n",
    "import db as db\n",
    "from utils_network import *\n",
    "from inference import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "056be149",
   "metadata": {},
   "outputs": [],
   "source": [
    "config_file = \"/data_nfs/og86asub/alternet-project/alternet/configs/MAGNet_NF.yaml\"\n",
    "import yaml\n",
    "\n",
    "with open(config_file, 'r') as f:\n",
    "    config = yaml.safe_load(f)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f120a43d",
   "metadata": {},
   "source": [
    "# Load necessary data and annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4e63cb94",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data_nfs/og86asub/alternet-project/alternet/src/alternet/db.py:32: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  digger = pd.read_csv(digger_path)\n"
     ]
    }
   ],
   "source": [
    "biomart_path = '/data_nfs/og86asub/alternet-project/alternet/data/biomart.txt'\n",
    "tf_list_path = '/data_nfs/og86asub/alternet-project/alternet/data/allTFs_hg38.txt'\n",
    "\n",
    "biomart = pd.read_csv(biomart_path, sep='\\t')\n",
    "tf_list = pd.read_csv(tf_list_path, sep='\\t', header = None)\n",
    "tf_list = map_tf_ids(tf_list, biomart)\n",
    "\n",
    "# create TF Database\n",
    "tf_database = db.create_transcipt_annotation_database(tf_list=tf_list, appris_path= config['appris'], digger_path=config['digger'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6fd4d8d6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3125cafc",
   "metadata": {},
   "outputs": [],
   "source": [
    "## LOAD gene and transcript data\n",
    "transcript_data = pd.read_csv(config['transcript_data'], index_col=0)\n",
    "\n",
    "## GET the gene ids associated with transcripts\n",
    "gene_ids = transcript_data['gene_id'].unique()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "23c39411",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>C00039</th>\n",
       "      <th>C00055</th>\n",
       "      <th>C00074</th>\n",
       "      <th>C00085</th>\n",
       "      <th>C00105</th>\n",
       "      <th>C00120</th>\n",
       "      <th>C00132</th>\n",
       "      <th>C00158</th>\n",
       "      <th>...</th>\n",
       "      <th>P01626</th>\n",
       "      <th>P01627</th>\n",
       "      <th>P01628</th>\n",
       "      <th>P01629</th>\n",
       "      <th>P01630</th>\n",
       "      <th>P01631</th>\n",
       "      <th>P01634</th>\n",
       "      <th>P01635</th>\n",
       "      <th>P01639</th>\n",
       "      <th>P01640</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000473530</td>\n",
       "      <td>ENSG00000244682</td>\n",
       "      <td>2.276150</td>\n",
       "      <td>2.040930</td>\n",
       "      <td>1.564230</td>\n",
       "      <td>2.629880</td>\n",
       "      <td>3.224940</td>\n",
       "      <td>0.081621</td>\n",
       "      <td>3.577190</td>\n",
       "      <td>3.275500</td>\n",
       "      <td>...</td>\n",
       "      <td>2.249030</td>\n",
       "      <td>0.754787</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.594620</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.031552</td>\n",
       "      <td>1.032080</td>\n",
       "      <td>2.658050</td>\n",
       "      <td>1.475990</td>\n",
       "      <td>0.012955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000467903</td>\n",
       "      <td>ENSG00000244682</td>\n",
       "      <td>2.167410</td>\n",
       "      <td>2.788660</td>\n",
       "      <td>3.829120</td>\n",
       "      <td>3.177140</td>\n",
       "      <td>4.211690</td>\n",
       "      <td>0.176553</td>\n",
       "      <td>6.959880</td>\n",
       "      <td>3.485350</td>\n",
       "      <td>...</td>\n",
       "      <td>4.777670</td>\n",
       "      <td>2.296700</td>\n",
       "      <td>0.256739</td>\n",
       "      <td>7.109500</td>\n",
       "      <td>0.226315</td>\n",
       "      <td>0.453683</td>\n",
       "      <td>1.436180</td>\n",
       "      <td>2.298990</td>\n",
       "      <td>3.807010</td>\n",
       "      <td>0.155549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000508651</td>\n",
       "      <td>ENSG00000244682</td>\n",
       "      <td>1.950990</td>\n",
       "      <td>2.569730</td>\n",
       "      <td>3.989870</td>\n",
       "      <td>3.134920</td>\n",
       "      <td>5.018950</td>\n",
       "      <td>0.110165</td>\n",
       "      <td>4.027630</td>\n",
       "      <td>1.973540</td>\n",
       "      <td>...</td>\n",
       "      <td>4.473280</td>\n",
       "      <td>1.117980</td>\n",
       "      <td>0.185469</td>\n",
       "      <td>3.784620</td>\n",
       "      <td>0.146539</td>\n",
       "      <td>0.085379</td>\n",
       "      <td>1.271050</td>\n",
       "      <td>1.629250</td>\n",
       "      <td>3.130670</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000626647</td>\n",
       "      <td>ENSG00000280938</td>\n",
       "      <td>0.049179</td>\n",
       "      <td>0.019637</td>\n",
       "      <td>0.017334</td>\n",
       "      <td>0.489663</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.105393</td>\n",
       "      <td>0.517216</td>\n",
       "      <td>0.029199</td>\n",
       "      <td>...</td>\n",
       "      <td>0.494979</td>\n",
       "      <td>3.469800</td>\n",
       "      <td>6.911700</td>\n",
       "      <td>1.803350</td>\n",
       "      <td>1.116030</td>\n",
       "      <td>0.850385</td>\n",
       "      <td>0.507154</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.293200</td>\n",
       "      <td>1.458430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000633377</td>\n",
       "      <td>ENSG00000282181</td>\n",
       "      <td>3.231750</td>\n",
       "      <td>1.134710</td>\n",
       "      <td>3.935840</td>\n",
       "      <td>0.548367</td>\n",
       "      <td>0.007815</td>\n",
       "      <td>0.319113</td>\n",
       "      <td>3.020600</td>\n",
       "      <td>1.936820</td>\n",
       "      <td>...</td>\n",
       "      <td>2.497850</td>\n",
       "      <td>0.987592</td>\n",
       "      <td>0.007706</td>\n",
       "      <td>0.507429</td>\n",
       "      <td>0.257754</td>\n",
       "      <td>0.544622</td>\n",
       "      <td>0.007459</td>\n",
       "      <td>1.090220</td>\n",
       "      <td>0.713997</td>\n",
       "      <td>1.031000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39310</th>\n",
       "      <td>ENST00000569097</td>\n",
       "      <td>ENSG00000197081</td>\n",
       "      <td>0.821758</td>\n",
       "      <td>0.377072</td>\n",
       "      <td>0.413925</td>\n",
       "      <td>0.907076</td>\n",
       "      <td>0.476571</td>\n",
       "      <td>0.098923</td>\n",
       "      <td>0.395161</td>\n",
       "      <td>0.379031</td>\n",
       "      <td>...</td>\n",
       "      <td>0.957061</td>\n",
       "      <td>0.418827</td>\n",
       "      <td>0.047140</td>\n",
       "      <td>0.320178</td>\n",
       "      <td>0.066159</td>\n",
       "      <td>0.132223</td>\n",
       "      <td>0.285653</td>\n",
       "      <td>0.631274</td>\n",
       "      <td>0.421508</td>\n",
       "      <td>0.062954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39311</th>\n",
       "      <td>ENST00000602699</td>\n",
       "      <td>ENSG00000225366</td>\n",
       "      <td>0.050290</td>\n",
       "      <td>0.014330</td>\n",
       "      <td>0.035171</td>\n",
       "      <td>0.041823</td>\n",
       "      <td>0.037060</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.065104</td>\n",
       "      <td>0.016308</td>\n",
       "      <td>...</td>\n",
       "      <td>0.044029</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.025864</td>\n",
       "      <td>0.024941</td>\n",
       "      <td>0.048181</td>\n",
       "      <td>0.038492</td>\n",
       "      <td>0.051400</td>\n",
       "      <td>0.017281</td>\n",
       "      <td>0.014494</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39312</th>\n",
       "      <td>ENST00000594408</td>\n",
       "      <td>ENSG00000182986</td>\n",
       "      <td>2.056060</td>\n",
       "      <td>1.068780</td>\n",
       "      <td>0.909087</td>\n",
       "      <td>1.211280</td>\n",
       "      <td>1.235250</td>\n",
       "      <td>0.022664</td>\n",
       "      <td>1.194140</td>\n",
       "      <td>1.571050</td>\n",
       "      <td>...</td>\n",
       "      <td>1.833670</td>\n",
       "      <td>0.982506</td>\n",
       "      <td>0.017389</td>\n",
       "      <td>0.592679</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.043223</td>\n",
       "      <td>0.683917</td>\n",
       "      <td>1.783930</td>\n",
       "      <td>1.038960</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39313</th>\n",
       "      <td>ENST00000510091</td>\n",
       "      <td>ENSG00000205482</td>\n",
       "      <td>0.397195</td>\n",
       "      <td>0.139399</td>\n",
       "      <td>0.137387</td>\n",
       "      <td>0.238294</td>\n",
       "      <td>0.456423</td>\n",
       "      <td>0.055239</td>\n",
       "      <td>0.127514</td>\n",
       "      <td>0.330749</td>\n",
       "      <td>...</td>\n",
       "      <td>1.001040</td>\n",
       "      <td>0.285062</td>\n",
       "      <td>0.049527</td>\n",
       "      <td>0.056683</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.040746</td>\n",
       "      <td>0.233459</td>\n",
       "      <td>0.148919</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39314</th>\n",
       "      <td>ENST00000618416</td>\n",
       "      <td>ENSG00000184616</td>\n",
       "      <td>0.769121</td>\n",
       "      <td>0.759989</td>\n",
       "      <td>0.346327</td>\n",
       "      <td>0.559473</td>\n",
       "      <td>0.161998</td>\n",
       "      <td>0.323150</td>\n",
       "      <td>1.094700</td>\n",
       "      <td>0.756638</td>\n",
       "      <td>...</td>\n",
       "      <td>0.476099</td>\n",
       "      <td>1.272690</td>\n",
       "      <td>0.116454</td>\n",
       "      <td>0.346888</td>\n",
       "      <td>0.063235</td>\n",
       "      <td>0.490208</td>\n",
       "      <td>0.384589</td>\n",
       "      <td>0.425519</td>\n",
       "      <td>0.308546</td>\n",
       "      <td>0.363469</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>39315 rows × 368 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         transcript_id          gene_id    C00039    C00055    C00074  \\\n",
       "0      ENST00000473530  ENSG00000244682  2.276150  2.040930  1.564230   \n",
       "1      ENST00000467903  ENSG00000244682  2.167410  2.788660  3.829120   \n",
       "2      ENST00000508651  ENSG00000244682  1.950990  2.569730  3.989870   \n",
       "3      ENST00000626647  ENSG00000280938  0.049179  0.019637  0.017334   \n",
       "4      ENST00000633377  ENSG00000282181  3.231750  1.134710  3.935840   \n",
       "...                ...              ...       ...       ...       ...   \n",
       "39310  ENST00000569097  ENSG00000197081  0.821758  0.377072  0.413925   \n",
       "39311  ENST00000602699  ENSG00000225366  0.050290  0.014330  0.035171   \n",
       "39312  ENST00000594408  ENSG00000182986  2.056060  1.068780  0.909087   \n",
       "39313  ENST00000510091  ENSG00000205482  0.397195  0.139399  0.137387   \n",
       "39314  ENST00000618416  ENSG00000184616  0.769121  0.759989  0.346327   \n",
       "\n",
       "         C00085    C00105    C00120    C00132    C00158  ...    P01626  \\\n",
       "0      2.629880  3.224940  0.081621  3.577190  3.275500  ...  2.249030   \n",
       "1      3.177140  4.211690  0.176553  6.959880  3.485350  ...  4.777670   \n",
       "2      3.134920  5.018950  0.110165  4.027630  1.973540  ...  4.473280   \n",
       "3      0.489663  0.000000  0.105393  0.517216  0.029199  ...  0.494979   \n",
       "4      0.548367  0.007815  0.319113  3.020600  1.936820  ...  2.497850   \n",
       "...         ...       ...       ...       ...       ...  ...       ...   \n",
       "39310  0.907076  0.476571  0.098923  0.395161  0.379031  ...  0.957061   \n",
       "39311  0.041823  0.037060  0.000000  0.065104  0.016308  ...  0.044029   \n",
       "39312  1.211280  1.235250  0.022664  1.194140  1.571050  ...  1.833670   \n",
       "39313  0.238294  0.456423  0.055239  0.127514  0.330749  ...  1.001040   \n",
       "39314  0.559473  0.161998  0.323150  1.094700  0.756638  ...  0.476099   \n",
       "\n",
       "         P01627    P01628    P01629    P01630    P01631    P01634    P01635  \\\n",
       "0      0.754787  0.000000  1.594620  0.000000  0.031552  1.032080  2.658050   \n",
       "1      2.296700  0.256739  7.109500  0.226315  0.453683  1.436180  2.298990   \n",
       "2      1.117980  0.185469  3.784620  0.146539  0.085379  1.271050  1.629250   \n",
       "3      3.469800  6.911700  1.803350  1.116030  0.850385  0.507154  0.000000   \n",
       "4      0.987592  0.007706  0.507429  0.257754  0.544622  0.007459  1.090220   \n",
       "...         ...       ...       ...       ...       ...       ...       ...   \n",
       "39310  0.418827  0.047140  0.320178  0.066159  0.132223  0.285653  0.631274   \n",
       "39311  0.000000  0.000000  0.025864  0.024941  0.048181  0.038492  0.051400   \n",
       "39312  0.982506  0.017389  0.592679  0.000000  0.043223  0.683917  1.783930   \n",
       "39313  0.285062  0.049527  0.056683  0.000000  0.000000  0.040746  0.233459   \n",
       "39314  1.272690  0.116454  0.346888  0.063235  0.490208  0.384589  0.425519   \n",
       "\n",
       "         P01639    P01640  \n",
       "0      1.475990  0.012955  \n",
       "1      3.807010  0.155549  \n",
       "2      3.130670  0.000000  \n",
       "3      1.293200  1.458430  \n",
       "4      0.713997  1.031000  \n",
       "...         ...       ...  \n",
       "39310  0.421508  0.062954  \n",
       "39311  0.017281  0.014494  \n",
       "39312  1.038960  0.000000  \n",
       "39313  0.148919  0.000000  \n",
       "39314  0.308546  0.363469  \n",
       "\n",
       "[39315 rows x 368 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcript_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "a77c583d",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_data = transcript_data.groupby('gene_id').sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f49829ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_data = gene_data.drop(columns={'transcript_id'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "70b0d1ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_data.index.name = 'gene_id'\n",
    "gene_data = gene_data.reset_index()\n",
    "gene_data = gene_data.iloc[0:40, ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "902ba019",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "## Subset the samples of interest\n",
    "sample_attributes = pd.read_csv(config['sample_attributes'])\n",
    "sample_attributes = sample_attributes.loc[:, ['sample_name', 'etiology']]\n",
    "samples = sample_attributes[sample_attributes['etiology'] == config['tissue']]\n",
    "samples = samples['sample_name'].tolist()\n",
    "\n",
    "\n",
    "## Get unified gene and transcript table\n",
    "gene_data = gene_data.loc[:, ['gene_id'] + samples ]\n",
    "transcript_data = transcript_data.loc[:,['gene_id', 'transcript_id'] + samples]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "32fdf04b",
   "metadata": {},
   "outputs": [],
   "source": [
    "tf_iso_counts, _ = separate_tf_genes(data=transcript_data, tf_list=tf_list, data_column='transcript_id', biomart_column='Transcript stable ID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "cf58253a",
   "metadata": {},
   "outputs": [],
   "source": [
    "tf_gene_counts, target_gene_counts = separate_tf_genes(data=gene_data, tf_list=tf_list, data_column='gene_id', biomart_column='Gene stable ID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "56eea2b8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Preparing data for inference: Separation into canonical and as-aware gene expression data\n"
     ]
    }
   ],
   "source": [
    "data_canonical, data_asware, target_gene_list = prepare_for_inference(tf_iso_counts, tf_gene_counts, target_gene_counts, transcript_column='transcript_id', gene_column='gene_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "dcb7e425",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>C00039</th>\n",
       "      <th>C00055</th>\n",
       "      <th>C00074</th>\n",
       "      <th>C00085</th>\n",
       "      <th>C00105</th>\n",
       "      <th>C00120</th>\n",
       "      <th>C00132</th>\n",
       "      <th>C00158</th>\n",
       "      <th>...</th>\n",
       "      <th>P01600</th>\n",
       "      <th>P01603</th>\n",
       "      <th>P01610</th>\n",
       "      <th>P01616</th>\n",
       "      <th>P01621</th>\n",
       "      <th>P01622</th>\n",
       "      <th>P01623</th>\n",
       "      <th>P01626</th>\n",
       "      <th>P01628</th>\n",
       "      <th>P01635</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>ENSG00000275746</td>\n",
       "      <td>ENST00000621843</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.921400</td>\n",
       "      <td>4.830320</td>\n",
       "      <td>4.321360</td>\n",
       "      <td>11.776600</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>11.941900</td>\n",
       "      <td>0.057060</td>\n",
       "      <td>...</td>\n",
       "      <td>3.274980</td>\n",
       "      <td>8.922970</td>\n",
       "      <td>12.219200</td>\n",
       "      <td>3.440300</td>\n",
       "      <td>0.035091</td>\n",
       "      <td>13.307500</td>\n",
       "      <td>12.556100</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.381808</td>\n",
       "      <td>3.594360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>ENSG00000280680</td>\n",
       "      <td>ENST00000629127</td>\n",
       "      <td>0.488933</td>\n",
       "      <td>0.006129</td>\n",
       "      <td>0.106360</td>\n",
       "      <td>0.269342</td>\n",
       "      <td>0.481044</td>\n",
       "      <td>0.208270</td>\n",
       "      <td>0.284457</td>\n",
       "      <td>0.230953</td>\n",
       "      <td>...</td>\n",
       "      <td>0.258495</td>\n",
       "      <td>0.109312</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.169916</td>\n",
       "      <td>0.193431</td>\n",
       "      <td>0.142123</td>\n",
       "      <td>0.291337</td>\n",
       "      <td>0.569886</td>\n",
       "      <td>0.167812</td>\n",
       "      <td>0.301905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>ENSG00000196458</td>\n",
       "      <td>ENST00000360187</td>\n",
       "      <td>1.556050</td>\n",
       "      <td>0.509873</td>\n",
       "      <td>0.559587</td>\n",
       "      <td>0.980809</td>\n",
       "      <td>0.613056</td>\n",
       "      <td>0.481815</td>\n",
       "      <td>0.643990</td>\n",
       "      <td>0.815159</td>\n",
       "      <td>...</td>\n",
       "      <td>0.672292</td>\n",
       "      <td>0.689361</td>\n",
       "      <td>0.727975</td>\n",
       "      <td>0.791609</td>\n",
       "      <td>0.408060</td>\n",
       "      <td>0.416713</td>\n",
       "      <td>0.931220</td>\n",
       "      <td>0.906321</td>\n",
       "      <td>0.456414</td>\n",
       "      <td>0.628137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>ENSG00000196458</td>\n",
       "      <td>ENST00000392321</td>\n",
       "      <td>4.305740</td>\n",
       "      <td>2.353420</td>\n",
       "      <td>1.963320</td>\n",
       "      <td>3.603160</td>\n",
       "      <td>2.979200</td>\n",
       "      <td>1.108800</td>\n",
       "      <td>2.318270</td>\n",
       "      <td>2.618990</td>\n",
       "      <td>...</td>\n",
       "      <td>2.885390</td>\n",
       "      <td>2.497590</td>\n",
       "      <td>2.936280</td>\n",
       "      <td>3.926700</td>\n",
       "      <td>0.836353</td>\n",
       "      <td>2.798880</td>\n",
       "      <td>3.036580</td>\n",
       "      <td>3.238990</td>\n",
       "      <td>0.998708</td>\n",
       "      <td>5.220130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>ENSG00000196458</td>\n",
       "      <td>ENST00000381215</td>\n",
       "      <td>1.360270</td>\n",
       "      <td>0.523519</td>\n",
       "      <td>0.718419</td>\n",
       "      <td>0.709772</td>\n",
       "      <td>0.507990</td>\n",
       "      <td>0.160236</td>\n",
       "      <td>1.083130</td>\n",
       "      <td>0.368735</td>\n",
       "      <td>...</td>\n",
       "      <td>0.433854</td>\n",
       "      <td>0.336460</td>\n",
       "      <td>0.596197</td>\n",
       "      <td>0.566359</td>\n",
       "      <td>0.175297</td>\n",
       "      <td>0.343090</td>\n",
       "      <td>0.530078</td>\n",
       "      <td>0.233906</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.563263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39234</th>\n",
       "      <td>ENSG00000282854</td>\n",
       "      <td>ENST00000634513</td>\n",
       "      <td>2.734940</td>\n",
       "      <td>2.158400</td>\n",
       "      <td>1.809580</td>\n",
       "      <td>1.739980</td>\n",
       "      <td>1.284910</td>\n",
       "      <td>0.136230</td>\n",
       "      <td>2.717190</td>\n",
       "      <td>1.536370</td>\n",
       "      <td>...</td>\n",
       "      <td>1.867850</td>\n",
       "      <td>1.626540</td>\n",
       "      <td>1.368840</td>\n",
       "      <td>1.454700</td>\n",
       "      <td>0.559354</td>\n",
       "      <td>1.852220</td>\n",
       "      <td>3.412960</td>\n",
       "      <td>3.710800</td>\n",
       "      <td>0.338195</td>\n",
       "      <td>1.194180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39235</th>\n",
       "      <td>ENSG00000282854</td>\n",
       "      <td>ENST00000634561</td>\n",
       "      <td>2.416900</td>\n",
       "      <td>2.097120</td>\n",
       "      <td>3.968760</td>\n",
       "      <td>1.306100</td>\n",
       "      <td>1.482140</td>\n",
       "      <td>0.064622</td>\n",
       "      <td>1.248550</td>\n",
       "      <td>1.534340</td>\n",
       "      <td>...</td>\n",
       "      <td>0.584410</td>\n",
       "      <td>2.874060</td>\n",
       "      <td>4.622200</td>\n",
       "      <td>0.932272</td>\n",
       "      <td>0.070226</td>\n",
       "      <td>3.783470</td>\n",
       "      <td>2.474310</td>\n",
       "      <td>2.291280</td>\n",
       "      <td>0.030406</td>\n",
       "      <td>1.310060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39236</th>\n",
       "      <td>ENSG00000282854</td>\n",
       "      <td>ENST00000635351</td>\n",
       "      <td>0.637849</td>\n",
       "      <td>0.098004</td>\n",
       "      <td>0.702740</td>\n",
       "      <td>0.536145</td>\n",
       "      <td>0.600501</td>\n",
       "      <td>0.413672</td>\n",
       "      <td>0.542548</td>\n",
       "      <td>0.522089</td>\n",
       "      <td>...</td>\n",
       "      <td>0.858770</td>\n",
       "      <td>0.219566</td>\n",
       "      <td>0.360110</td>\n",
       "      <td>0.286954</td>\n",
       "      <td>0.063341</td>\n",
       "      <td>0.107373</td>\n",
       "      <td>0.438501</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.338502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39237</th>\n",
       "      <td>ENSG00000282899</td>\n",
       "      <td>ENST00000634683</td>\n",
       "      <td>0.015705</td>\n",
       "      <td>0.018329</td>\n",
       "      <td>0.048345</td>\n",
       "      <td>0.009561</td>\n",
       "      <td>0.085228</td>\n",
       "      <td>0.021931</td>\n",
       "      <td>0.050436</td>\n",
       "      <td>0.031760</td>\n",
       "      <td>...</td>\n",
       "      <td>0.050812</td>\n",
       "      <td>0.013758</td>\n",
       "      <td>0.011409</td>\n",
       "      <td>0.023511</td>\n",
       "      <td>0.052863</td>\n",
       "      <td>0.020997</td>\n",
       "      <td>0.038926</td>\n",
       "      <td>0.049232</td>\n",
       "      <td>0.006934</td>\n",
       "      <td>0.011964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39312</th>\n",
       "      <td>ENSG00000182986</td>\n",
       "      <td>ENST00000594408</td>\n",
       "      <td>2.056060</td>\n",
       "      <td>1.068780</td>\n",
       "      <td>0.909087</td>\n",
       "      <td>1.211280</td>\n",
       "      <td>1.235250</td>\n",
       "      <td>0.022664</td>\n",
       "      <td>1.194140</td>\n",
       "      <td>1.571050</td>\n",
       "      <td>...</td>\n",
       "      <td>2.038440</td>\n",
       "      <td>0.777621</td>\n",
       "      <td>0.761086</td>\n",
       "      <td>1.157570</td>\n",
       "      <td>0.064273</td>\n",
       "      <td>0.759715</td>\n",
       "      <td>1.245030</td>\n",
       "      <td>1.833670</td>\n",
       "      <td>0.017389</td>\n",
       "      <td>1.783930</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3489 rows × 168 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene_id    transcript_id    C00039    C00055    C00074  \\\n",
       "69     ENSG00000275746  ENST00000621843  0.000000  1.921400  4.830320   \n",
       "71     ENSG00000280680  ENST00000629127  0.488933  0.006129  0.106360   \n",
       "141    ENSG00000196458  ENST00000360187  1.556050  0.509873  0.559587   \n",
       "142    ENSG00000196458  ENST00000392321  4.305740  2.353420  1.963320   \n",
       "143    ENSG00000196458  ENST00000381215  1.360270  0.523519  0.718419   \n",
       "...                ...              ...       ...       ...       ...   \n",
       "39234  ENSG00000282854  ENST00000634513  2.734940  2.158400  1.809580   \n",
       "39235  ENSG00000282854  ENST00000634561  2.416900  2.097120  3.968760   \n",
       "39236  ENSG00000282854  ENST00000635351  0.637849  0.098004  0.702740   \n",
       "39237  ENSG00000282899  ENST00000634683  0.015705  0.018329  0.048345   \n",
       "39312  ENSG00000182986  ENST00000594408  2.056060  1.068780  0.909087   \n",
       "\n",
       "         C00085     C00105    C00120     C00132    C00158  ...    P01600  \\\n",
       "69     4.321360  11.776600  0.000000  11.941900  0.057060  ...  3.274980   \n",
       "71     0.269342   0.481044  0.208270   0.284457  0.230953  ...  0.258495   \n",
       "141    0.980809   0.613056  0.481815   0.643990  0.815159  ...  0.672292   \n",
       "142    3.603160   2.979200  1.108800   2.318270  2.618990  ...  2.885390   \n",
       "143    0.709772   0.507990  0.160236   1.083130  0.368735  ...  0.433854   \n",
       "...         ...        ...       ...        ...       ...  ...       ...   \n",
       "39234  1.739980   1.284910  0.136230   2.717190  1.536370  ...  1.867850   \n",
       "39235  1.306100   1.482140  0.064622   1.248550  1.534340  ...  0.584410   \n",
       "39236  0.536145   0.600501  0.413672   0.542548  0.522089  ...  0.858770   \n",
       "39237  0.009561   0.085228  0.021931   0.050436  0.031760  ...  0.050812   \n",
       "39312  1.211280   1.235250  0.022664   1.194140  1.571050  ...  2.038440   \n",
       "\n",
       "         P01603     P01610    P01616    P01621     P01622     P01623  \\\n",
       "69     8.922970  12.219200  3.440300  0.035091  13.307500  12.556100   \n",
       "71     0.109312   0.000000  0.169916  0.193431   0.142123   0.291337   \n",
       "141    0.689361   0.727975  0.791609  0.408060   0.416713   0.931220   \n",
       "142    2.497590   2.936280  3.926700  0.836353   2.798880   3.036580   \n",
       "143    0.336460   0.596197  0.566359  0.175297   0.343090   0.530078   \n",
       "...         ...        ...       ...       ...        ...        ...   \n",
       "39234  1.626540   1.368840  1.454700  0.559354   1.852220   3.412960   \n",
       "39235  2.874060   4.622200  0.932272  0.070226   3.783470   2.474310   \n",
       "39236  0.219566   0.360110  0.286954  0.063341   0.107373   0.438501   \n",
       "39237  0.013758   0.011409  0.023511  0.052863   0.020997   0.038926   \n",
       "39312  0.777621   0.761086  1.157570  0.064273   0.759715   1.245030   \n",
       "\n",
       "         P01626    P01628    P01635  \n",
       "69     0.000000  0.381808  3.594360  \n",
       "71     0.569886  0.167812  0.301905  \n",
       "141    0.906321  0.456414  0.628137  \n",
       "142    3.238990  0.998708  5.220130  \n",
       "143    0.233906  0.000000  0.563263  \n",
       "...         ...       ...       ...  \n",
       "39234  3.710800  0.338195  1.194180  \n",
       "39235  2.291280  0.030406  1.310060  \n",
       "39236  0.000000  0.000000  0.338502  \n",
       "39237  0.049232  0.006934  0.011964  \n",
       "39312  1.833670  0.017389  1.783930  \n",
       "\n",
       "[3489 rows x 168 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tf_iso_counts\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "786317af",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>C00039</th>\n",
       "      <th>C00055</th>\n",
       "      <th>C00074</th>\n",
       "      <th>C00085</th>\n",
       "      <th>C00105</th>\n",
       "      <th>C00120</th>\n",
       "      <th>C00132</th>\n",
       "      <th>C00158</th>\n",
       "      <th>C00168</th>\n",
       "      <th>C00186</th>\n",
       "      <th>...</th>\n",
       "      <th>P01600</th>\n",
       "      <th>P01603</th>\n",
       "      <th>P01610</th>\n",
       "      <th>P01616</th>\n",
       "      <th>P01621</th>\n",
       "      <th>P01622</th>\n",
       "      <th>P01623</th>\n",
       "      <th>P01626</th>\n",
       "      <th>P01628</th>\n",
       "      <th>P01635</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENST00000621843</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.921400</td>\n",
       "      <td>4.830320</td>\n",
       "      <td>4.321360</td>\n",
       "      <td>11.776600</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>11.941900</td>\n",
       "      <td>0.057060</td>\n",
       "      <td>3.526590</td>\n",
       "      <td>0.170855</td>\n",
       "      <td>...</td>\n",
       "      <td>3.274980</td>\n",
       "      <td>8.922970</td>\n",
       "      <td>12.219200</td>\n",
       "      <td>3.440300</td>\n",
       "      <td>0.035091</td>\n",
       "      <td>13.307500</td>\n",
       "      <td>12.556100</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.381808</td>\n",
       "      <td>3.594360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000629127</th>\n",
       "      <td>0.488933</td>\n",
       "      <td>0.006129</td>\n",
       "      <td>0.106360</td>\n",
       "      <td>0.269342</td>\n",
       "      <td>0.481044</td>\n",
       "      <td>0.208270</td>\n",
       "      <td>0.284457</td>\n",
       "      <td>0.230953</td>\n",
       "      <td>0.408554</td>\n",
       "      <td>0.173717</td>\n",
       "      <td>...</td>\n",
       "      <td>0.258495</td>\n",
       "      <td>0.109312</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.169916</td>\n",
       "      <td>0.193431</td>\n",
       "      <td>0.142123</td>\n",
       "      <td>0.291337</td>\n",
       "      <td>0.569886</td>\n",
       "      <td>0.167812</td>\n",
       "      <td>0.301905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000360187</th>\n",
       "      <td>1.556050</td>\n",
       "      <td>0.509873</td>\n",
       "      <td>0.559587</td>\n",
       "      <td>0.980809</td>\n",
       "      <td>0.613056</td>\n",
       "      <td>0.481815</td>\n",
       "      <td>0.643990</td>\n",
       "      <td>0.815159</td>\n",
       "      <td>0.685427</td>\n",
       "      <td>0.802393</td>\n",
       "      <td>...</td>\n",
       "      <td>0.672292</td>\n",
       "      <td>0.689361</td>\n",
       "      <td>0.727975</td>\n",
       "      <td>0.791609</td>\n",
       "      <td>0.408060</td>\n",
       "      <td>0.416713</td>\n",
       "      <td>0.931220</td>\n",
       "      <td>0.906321</td>\n",
       "      <td>0.456414</td>\n",
       "      <td>0.628137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000392321</th>\n",
       "      <td>4.305740</td>\n",
       "      <td>2.353420</td>\n",
       "      <td>1.963320</td>\n",
       "      <td>3.603160</td>\n",
       "      <td>2.979200</td>\n",
       "      <td>1.108800</td>\n",
       "      <td>2.318270</td>\n",
       "      <td>2.618990</td>\n",
       "      <td>1.960630</td>\n",
       "      <td>0.120433</td>\n",
       "      <td>...</td>\n",
       "      <td>2.885390</td>\n",
       "      <td>2.497590</td>\n",
       "      <td>2.936280</td>\n",
       "      <td>3.926700</td>\n",
       "      <td>0.836353</td>\n",
       "      <td>2.798880</td>\n",
       "      <td>3.036580</td>\n",
       "      <td>3.238990</td>\n",
       "      <td>0.998708</td>\n",
       "      <td>5.220130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000381215</th>\n",
       "      <td>1.360270</td>\n",
       "      <td>0.523519</td>\n",
       "      <td>0.718419</td>\n",
       "      <td>0.709772</td>\n",
       "      <td>0.507990</td>\n",
       "      <td>0.160236</td>\n",
       "      <td>1.083130</td>\n",
       "      <td>0.368735</td>\n",
       "      <td>0.229089</td>\n",
       "      <td>0.018763</td>\n",
       "      <td>...</td>\n",
       "      <td>0.433854</td>\n",
       "      <td>0.336460</td>\n",
       "      <td>0.596197</td>\n",
       "      <td>0.566359</td>\n",
       "      <td>0.175297</td>\n",
       "      <td>0.343090</td>\n",
       "      <td>0.530078</td>\n",
       "      <td>0.233906</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.563263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000003436</th>\n",
       "      <td>28.490930</td>\n",
       "      <td>18.440540</td>\n",
       "      <td>15.969181</td>\n",
       "      <td>21.885337</td>\n",
       "      <td>22.476400</td>\n",
       "      <td>10.506746</td>\n",
       "      <td>27.181098</td>\n",
       "      <td>20.534197</td>\n",
       "      <td>12.225895</td>\n",
       "      <td>10.116812</td>\n",
       "      <td>...</td>\n",
       "      <td>15.251027</td>\n",
       "      <td>9.683091</td>\n",
       "      <td>13.895921</td>\n",
       "      <td>14.239894</td>\n",
       "      <td>18.224117</td>\n",
       "      <td>11.555473</td>\n",
       "      <td>14.553964</td>\n",
       "      <td>21.890835</td>\n",
       "      <td>16.586082</td>\n",
       "      <td>19.743486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000003509</th>\n",
       "      <td>21.012959</td>\n",
       "      <td>18.172372</td>\n",
       "      <td>22.750115</td>\n",
       "      <td>16.056193</td>\n",
       "      <td>18.383836</td>\n",
       "      <td>15.862320</td>\n",
       "      <td>19.250591</td>\n",
       "      <td>17.897856</td>\n",
       "      <td>18.642779</td>\n",
       "      <td>8.065930</td>\n",
       "      <td>...</td>\n",
       "      <td>20.107785</td>\n",
       "      <td>11.898181</td>\n",
       "      <td>17.237519</td>\n",
       "      <td>17.251907</td>\n",
       "      <td>8.558163</td>\n",
       "      <td>13.976136</td>\n",
       "      <td>13.146203</td>\n",
       "      <td>17.406819</td>\n",
       "      <td>8.288946</td>\n",
       "      <td>15.996116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000003756</th>\n",
       "      <td>54.936501</td>\n",
       "      <td>26.129274</td>\n",
       "      <td>38.328458</td>\n",
       "      <td>45.693790</td>\n",
       "      <td>42.495645</td>\n",
       "      <td>28.709693</td>\n",
       "      <td>37.254217</td>\n",
       "      <td>35.839398</td>\n",
       "      <td>28.977764</td>\n",
       "      <td>17.107432</td>\n",
       "      <td>...</td>\n",
       "      <td>43.215013</td>\n",
       "      <td>23.500684</td>\n",
       "      <td>34.498117</td>\n",
       "      <td>34.216680</td>\n",
       "      <td>21.748461</td>\n",
       "      <td>20.232026</td>\n",
       "      <td>32.258981</td>\n",
       "      <td>51.464930</td>\n",
       "      <td>19.335575</td>\n",
       "      <td>42.815704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000003987</th>\n",
       "      <td>0.673885</td>\n",
       "      <td>0.513828</td>\n",
       "      <td>0.733176</td>\n",
       "      <td>0.730261</td>\n",
       "      <td>0.811276</td>\n",
       "      <td>0.212408</td>\n",
       "      <td>0.698341</td>\n",
       "      <td>0.470540</td>\n",
       "      <td>0.636159</td>\n",
       "      <td>0.271111</td>\n",
       "      <td>...</td>\n",
       "      <td>0.540007</td>\n",
       "      <td>0.357516</td>\n",
       "      <td>0.366972</td>\n",
       "      <td>0.680045</td>\n",
       "      <td>0.206822</td>\n",
       "      <td>0.313873</td>\n",
       "      <td>0.557605</td>\n",
       "      <td>0.970646</td>\n",
       "      <td>0.398043</td>\n",
       "      <td>0.640866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000003989</th>\n",
       "      <td>19.475052</td>\n",
       "      <td>10.481757</td>\n",
       "      <td>17.166040</td>\n",
       "      <td>13.646568</td>\n",
       "      <td>21.286419</td>\n",
       "      <td>6.557228</td>\n",
       "      <td>12.490144</td>\n",
       "      <td>14.942467</td>\n",
       "      <td>11.024453</td>\n",
       "      <td>1.786017</td>\n",
       "      <td>...</td>\n",
       "      <td>7.917652</td>\n",
       "      <td>5.952005</td>\n",
       "      <td>10.878596</td>\n",
       "      <td>12.334331</td>\n",
       "      <td>2.804582</td>\n",
       "      <td>6.817518</td>\n",
       "      <td>4.187878</td>\n",
       "      <td>11.037871</td>\n",
       "      <td>3.648987</td>\n",
       "      <td>10.234291</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3526 rows × 166 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    C00039     C00055     C00074     C00085     C00105  \\\n",
       "gene_id                                                                  \n",
       "ENST00000621843   0.000000   1.921400   4.830320   4.321360  11.776600   \n",
       "ENST00000629127   0.488933   0.006129   0.106360   0.269342   0.481044   \n",
       "ENST00000360187   1.556050   0.509873   0.559587   0.980809   0.613056   \n",
       "ENST00000392321   4.305740   2.353420   1.963320   3.603160   2.979200   \n",
       "ENST00000381215   1.360270   0.523519   0.718419   0.709772   0.507990   \n",
       "...                    ...        ...        ...        ...        ...   \n",
       "ENSG00000003436  28.490930  18.440540  15.969181  21.885337  22.476400   \n",
       "ENSG00000003509  21.012959  18.172372  22.750115  16.056193  18.383836   \n",
       "ENSG00000003756  54.936501  26.129274  38.328458  45.693790  42.495645   \n",
       "ENSG00000003987   0.673885   0.513828   0.733176   0.730261   0.811276   \n",
       "ENSG00000003989  19.475052  10.481757  17.166040  13.646568  21.286419   \n",
       "\n",
       "                    C00120     C00132     C00158     C00168     C00186  ...  \\\n",
       "gene_id                                                                 ...   \n",
       "ENST00000621843   0.000000  11.941900   0.057060   3.526590   0.170855  ...   \n",
       "ENST00000629127   0.208270   0.284457   0.230953   0.408554   0.173717  ...   \n",
       "ENST00000360187   0.481815   0.643990   0.815159   0.685427   0.802393  ...   \n",
       "ENST00000392321   1.108800   2.318270   2.618990   1.960630   0.120433  ...   \n",
       "ENST00000381215   0.160236   1.083130   0.368735   0.229089   0.018763  ...   \n",
       "...                    ...        ...        ...        ...        ...  ...   \n",
       "ENSG00000003436  10.506746  27.181098  20.534197  12.225895  10.116812  ...   \n",
       "ENSG00000003509  15.862320  19.250591  17.897856  18.642779   8.065930  ...   \n",
       "ENSG00000003756  28.709693  37.254217  35.839398  28.977764  17.107432  ...   \n",
       "ENSG00000003987   0.212408   0.698341   0.470540   0.636159   0.271111  ...   \n",
       "ENSG00000003989   6.557228  12.490144  14.942467  11.024453   1.786017  ...   \n",
       "\n",
       "                    P01600     P01603     P01610     P01616     P01621  \\\n",
       "gene_id                                                                  \n",
       "ENST00000621843   3.274980   8.922970  12.219200   3.440300   0.035091   \n",
       "ENST00000629127   0.258495   0.109312   0.000000   0.169916   0.193431   \n",
       "ENST00000360187   0.672292   0.689361   0.727975   0.791609   0.408060   \n",
       "ENST00000392321   2.885390   2.497590   2.936280   3.926700   0.836353   \n",
       "ENST00000381215   0.433854   0.336460   0.596197   0.566359   0.175297   \n",
       "...                    ...        ...        ...        ...        ...   \n",
       "ENSG00000003436  15.251027   9.683091  13.895921  14.239894  18.224117   \n",
       "ENSG00000003509  20.107785  11.898181  17.237519  17.251907   8.558163   \n",
       "ENSG00000003756  43.215013  23.500684  34.498117  34.216680  21.748461   \n",
       "ENSG00000003987   0.540007   0.357516   0.366972   0.680045   0.206822   \n",
       "ENSG00000003989   7.917652   5.952005  10.878596  12.334331   2.804582   \n",
       "\n",
       "                    P01622     P01623     P01626     P01628     P01635  \n",
       "gene_id                                                                 \n",
       "ENST00000621843  13.307500  12.556100   0.000000   0.381808   3.594360  \n",
       "ENST00000629127   0.142123   0.291337   0.569886   0.167812   0.301905  \n",
       "ENST00000360187   0.416713   0.931220   0.906321   0.456414   0.628137  \n",
       "ENST00000392321   2.798880   3.036580   3.238990   0.998708   5.220130  \n",
       "ENST00000381215   0.343090   0.530078   0.233906   0.000000   0.563263  \n",
       "...                    ...        ...        ...        ...        ...  \n",
       "ENSG00000003436  11.555473  14.553964  21.890835  16.586082  19.743486  \n",
       "ENSG00000003509  13.976136  13.146203  17.406819   8.288946  15.996116  \n",
       "ENSG00000003756  20.232026  32.258981  51.464930  19.335575  42.815704  \n",
       "ENSG00000003987   0.313873   0.557605   0.970646   0.398043   0.640866  \n",
       "ENSG00000003989   6.817518   4.187878  11.037871   3.648987  10.234291  \n",
       "\n",
       "[3526 rows x 166 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_asware"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "8a4e35eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "            gene_id     C00039     C00055     C00074    C00085     C00105  \\\n",
      "9   ENSG00000001167   7.682690   4.104230   4.853400  4.317800   5.404930   \n",
      "12  ENSG00000001497  15.656040   7.594953  11.370840  9.514930   8.113490   \n",
      "19  ENSG00000002330   6.494686  11.965629  16.796471  5.533693  13.560036   \n",
      "\n",
      "       C00120     C00132     C00158    C00168  ...     P01603     P01610  \\\n",
      "9    0.896633   3.439820   4.188700   2.40911  ...   3.054800   4.432720   \n",
      "12   6.973793  10.199250   9.995520  12.91687  ...   6.606053   8.879420   \n",
      "19  36.562488  10.916904  11.033318  18.67241  ...  28.632750  62.501836   \n",
      "\n",
      "      P01616     P01621     P01622     P01623     P01626     P01628    P01635  \\\n",
      "9   4.571550   0.865780   3.645890   2.541630   4.436620   0.818385  6.867520   \n",
      "12  7.067700   7.164017   5.129949  11.200990  12.328330   6.447379  8.749470   \n",
      "19  1.875371  65.369523  46.345562  12.156581  12.816146  64.315560  4.524684   \n",
      "\n",
      "    sum_gene_expression  \n",
      "9            598.263609  \n",
      "12          1639.393565  \n",
      "19          4108.186393  \n",
      "\n",
      "[3 rows x 168 columns]\n"
     ]
    }
   ],
   "source": [
    "# get isoform categories\n",
    "isoform_categories = isoform_categorization(tf_iso_counts, tf_gene_counts, threshold_dominance=80)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "a9fd1ab3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>C00039</th>\n",
       "      <th>C00055</th>\n",
       "      <th>C00074</th>\n",
       "      <th>C00085</th>\n",
       "      <th>C00105</th>\n",
       "      <th>C00120</th>\n",
       "      <th>C00132</th>\n",
       "      <th>C00158</th>\n",
       "      <th>C00168</th>\n",
       "      <th>...</th>\n",
       "      <th>P01600</th>\n",
       "      <th>P01603</th>\n",
       "      <th>P01610</th>\n",
       "      <th>P01616</th>\n",
       "      <th>P01621</th>\n",
       "      <th>P01622</th>\n",
       "      <th>P01623</th>\n",
       "      <th>P01626</th>\n",
       "      <th>P01628</th>\n",
       "      <th>P01635</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSG00000001167</td>\n",
       "      <td>7.682690</td>\n",
       "      <td>4.104230</td>\n",
       "      <td>4.853400</td>\n",
       "      <td>4.317800</td>\n",
       "      <td>5.404930</td>\n",
       "      <td>0.896633</td>\n",
       "      <td>3.439820</td>\n",
       "      <td>4.188700</td>\n",
       "      <td>2.40911</td>\n",
       "      <td>...</td>\n",
       "      <td>4.120600</td>\n",
       "      <td>3.054800</td>\n",
       "      <td>4.432720</td>\n",
       "      <td>4.571550</td>\n",
       "      <td>0.865780</td>\n",
       "      <td>3.645890</td>\n",
       "      <td>2.541630</td>\n",
       "      <td>4.436620</td>\n",
       "      <td>0.818385</td>\n",
       "      <td>6.867520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENSG00000001497</td>\n",
       "      <td>15.656040</td>\n",
       "      <td>7.594953</td>\n",
       "      <td>11.370840</td>\n",
       "      <td>9.514930</td>\n",
       "      <td>8.113490</td>\n",
       "      <td>6.973793</td>\n",
       "      <td>10.199250</td>\n",
       "      <td>9.995520</td>\n",
       "      <td>12.91687</td>\n",
       "      <td>...</td>\n",
       "      <td>10.265280</td>\n",
       "      <td>6.606053</td>\n",
       "      <td>8.879420</td>\n",
       "      <td>7.067700</td>\n",
       "      <td>7.164017</td>\n",
       "      <td>5.129949</td>\n",
       "      <td>11.200990</td>\n",
       "      <td>12.328330</td>\n",
       "      <td>6.447379</td>\n",
       "      <td>8.749470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENSG00000002330</td>\n",
       "      <td>6.494686</td>\n",
       "      <td>11.965629</td>\n",
       "      <td>16.796471</td>\n",
       "      <td>5.533693</td>\n",
       "      <td>13.560036</td>\n",
       "      <td>36.562488</td>\n",
       "      <td>10.916904</td>\n",
       "      <td>11.033318</td>\n",
       "      <td>18.67241</td>\n",
       "      <td>...</td>\n",
       "      <td>4.760836</td>\n",
       "      <td>28.632750</td>\n",
       "      <td>62.501836</td>\n",
       "      <td>1.875371</td>\n",
       "      <td>65.369523</td>\n",
       "      <td>46.345562</td>\n",
       "      <td>12.156581</td>\n",
       "      <td>12.816146</td>\n",
       "      <td>64.315560</td>\n",
       "      <td>4.524684</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 167 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene_id     C00039     C00055     C00074    C00085     C00105  \\\n",
       "9   ENSG00000001167   7.682690   4.104230   4.853400  4.317800   5.404930   \n",
       "12  ENSG00000001497  15.656040   7.594953  11.370840  9.514930   8.113490   \n",
       "19  ENSG00000002330   6.494686  11.965629  16.796471  5.533693  13.560036   \n",
       "\n",
       "       C00120     C00132     C00158    C00168  ...     P01600     P01603  \\\n",
       "9    0.896633   3.439820   4.188700   2.40911  ...   4.120600   3.054800   \n",
       "12   6.973793  10.199250   9.995520  12.91687  ...  10.265280   6.606053   \n",
       "19  36.562488  10.916904  11.033318  18.67241  ...   4.760836  28.632750   \n",
       "\n",
       "       P01610    P01616     P01621     P01622     P01623     P01626  \\\n",
       "9    4.432720  4.571550   0.865780   3.645890   2.541630   4.436620   \n",
       "12   8.879420  7.067700   7.164017   5.129949  11.200990  12.328330   \n",
       "19  62.501836  1.875371  65.369523  46.345562  12.156581  12.816146   \n",
       "\n",
       "       P01628    P01635  \n",
       "9    0.818385  6.867520  \n",
       "12   6.447379  8.749470  \n",
       "19  64.315560  4.524684  \n",
       "\n",
       "[3 rows x 167 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tf_gene_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "7933898d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>sum_transcript_expression</th>\n",
       "      <th>sum_gene_expression</th>\n",
       "      <th>percentage</th>\n",
       "      <th>max_percentage</th>\n",
       "      <th>min_percentage</th>\n",
       "      <th>std_percentage</th>\n",
       "      <th>isoform_category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000394532</td>\n",
       "      <td>ENSG00000002330</td>\n",
       "      <td>168.174</td>\n",
       "      <td>4108.186</td>\n",
       "      <td>4.094</td>\n",
       "      <td>86.225</td>\n",
       "      <td>2.390</td>\n",
       "      <td>40.867</td>\n",
       "      <td>non-dominant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000309032</td>\n",
       "      <td>ENSG00000002330</td>\n",
       "      <td>3542.295</td>\n",
       "      <td>4108.186</td>\n",
       "      <td>86.225</td>\n",
       "      <td>86.225</td>\n",
       "      <td>2.390</td>\n",
       "      <td>40.867</td>\n",
       "      <td>dominant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000492141</td>\n",
       "      <td>ENSG00000002330</td>\n",
       "      <td>98.202</td>\n",
       "      <td>4108.186</td>\n",
       "      <td>2.390</td>\n",
       "      <td>86.225</td>\n",
       "      <td>2.390</td>\n",
       "      <td>40.867</td>\n",
       "      <td>non-dominant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000544271</td>\n",
       "      <td>ENSG00000002330</td>\n",
       "      <td>299.516</td>\n",
       "      <td>4108.186</td>\n",
       "      <td>7.291</td>\n",
       "      <td>86.225</td>\n",
       "      <td>2.390</td>\n",
       "      <td>40.867</td>\n",
       "      <td>non-dominant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000374807</td>\n",
       "      <td>ENSG00000001497</td>\n",
       "      <td>309.502</td>\n",
       "      <td>1639.394</td>\n",
       "      <td>18.879</td>\n",
       "      <td>34.838</td>\n",
       "      <td>12.950</td>\n",
       "      <td>10.784</td>\n",
       "      <td>balanced</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000374811</td>\n",
       "      <td>ENSG00000001497</td>\n",
       "      <td>546.445</td>\n",
       "      <td>1639.394</td>\n",
       "      <td>33.332</td>\n",
       "      <td>34.838</td>\n",
       "      <td>12.950</td>\n",
       "      <td>10.784</td>\n",
       "      <td>balanced</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000484069</td>\n",
       "      <td>ENSG00000001497</td>\n",
       "      <td>212.309</td>\n",
       "      <td>1639.394</td>\n",
       "      <td>12.950</td>\n",
       "      <td>34.838</td>\n",
       "      <td>12.950</td>\n",
       "      <td>10.784</td>\n",
       "      <td>balanced</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000469091</td>\n",
       "      <td>ENSG00000001497</td>\n",
       "      <td>571.137</td>\n",
       "      <td>1639.394</td>\n",
       "      <td>34.838</td>\n",
       "      <td>34.838</td>\n",
       "      <td>12.950</td>\n",
       "      <td>10.784</td>\n",
       "      <td>balanced</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000341376</td>\n",
       "      <td>ENSG00000001167</td>\n",
       "      <td>598.264</td>\n",
       "      <td>598.264</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dominant</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     transcript_id          gene_id  sum_transcript_expression  \\\n",
       "0  ENST00000394532  ENSG00000002330                    168.174   \n",
       "1  ENST00000309032  ENSG00000002330                   3542.295   \n",
       "2  ENST00000492141  ENSG00000002330                     98.202   \n",
       "3  ENST00000544271  ENSG00000002330                    299.516   \n",
       "4  ENST00000374807  ENSG00000001497                    309.502   \n",
       "5  ENST00000374811  ENSG00000001497                    546.445   \n",
       "6  ENST00000484069  ENSG00000001497                    212.309   \n",
       "7  ENST00000469091  ENSG00000001497                    571.137   \n",
       "8  ENST00000341376  ENSG00000001167                    598.264   \n",
       "\n",
       "   sum_gene_expression  percentage  max_percentage  min_percentage  \\\n",
       "0             4108.186       4.094          86.225           2.390   \n",
       "1             4108.186      86.225          86.225           2.390   \n",
       "2             4108.186       2.390          86.225           2.390   \n",
       "3             4108.186       7.291          86.225           2.390   \n",
       "4             1639.394      18.879          34.838          12.950   \n",
       "5             1639.394      33.332          34.838          12.950   \n",
       "6             1639.394      12.950          34.838          12.950   \n",
       "7             1639.394      34.838          34.838          12.950   \n",
       "8              598.264     100.000         100.000         100.000   \n",
       "\n",
       "   std_percentage isoform_category  \n",
       "0          40.867     non-dominant  \n",
       "1          40.867         dominant  \n",
       "2          40.867     non-dominant  \n",
       "3          40.867     non-dominant  \n",
       "4          10.784         balanced  \n",
       "5          10.784         balanced  \n",
       "6          10.784         balanced  \n",
       "7          10.784         balanced  \n",
       "8             NaN         dominant  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.set_option('display.float_format', lambda x: '%.3f' % x)\n",
    "isoform_categories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "da84dd89",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting inference ...\n",
      "Computing network\n",
      "Computing network\n",
      "Computing network\n",
      "Computing network\n",
      "Inference complete\n",
      "Aggregate results\n",
      "Going through edges of each df ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 38846/38846 [00:00<00:00, 41915.22it/s]\n",
      "100%|██████████| 38197/38197 [00:00<00:00, 41793.96it/s]\n",
      "100%|██████████| 58624/58624 [00:00<00:00, 1977345.58it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Going through edges of each df ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 111/111 [00:00<00:00, 34491.61it/s]\n",
      "100%|██████████| 111/111 [00:00<00:00, 39278.47it/s]\n",
      "100%|██████████| 111/111 [00:00<00:00, 1293243.73it/s]\n"
     ]
    }
   ],
   "source": [
    "as_aware_grn, canonical_grn = inference(config, 2, tf_list=tf_list, data_canonical=data_canonical, data_asware=data_asware, target_gene_list=target_gene_list, aggregate=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "bd38ec46",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>frequency</th>\n",
       "      <th>mean_importance</th>\n",
       "      <th>median_importance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000001167</td>\n",
       "      <td>ENSG00000003056</td>\n",
       "      <td>2</td>\n",
       "      <td>173.096</td>\n",
       "      <td>178.614</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000001167</td>\n",
       "      <td>ENSG00000001629</td>\n",
       "      <td>2</td>\n",
       "      <td>166.671</td>\n",
       "      <td>168.557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000001167</td>\n",
       "      <td>ENSG00000001561</td>\n",
       "      <td>2</td>\n",
       "      <td>132.995</td>\n",
       "      <td>137.291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000001167</td>\n",
       "      <td>ENSG00000001460</td>\n",
       "      <td>2</td>\n",
       "      <td>125.436</td>\n",
       "      <td>128.146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000001167</td>\n",
       "      <td>ENSG00000003393</td>\n",
       "      <td>2</td>\n",
       "      <td>131.500</td>\n",
       "      <td>141.749</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>ENSG00000002330</td>\n",
       "      <td>ENSG00000001036</td>\n",
       "      <td>2</td>\n",
       "      <td>11.414</td>\n",
       "      <td>12.249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>ENSG00000001497</td>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>2</td>\n",
       "      <td>8.448</td>\n",
       "      <td>8.633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>ENSG00000002330</td>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>2</td>\n",
       "      <td>7.225</td>\n",
       "      <td>8.388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>ENSG00000001497</td>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>2</td>\n",
       "      <td>17.364</td>\n",
       "      <td>26.507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>ENSG00000001167</td>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>2</td>\n",
       "      <td>10.929</td>\n",
       "      <td>17.483</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>111 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              source           target  frequency  mean_importance  \\\n",
       "0    ENSG00000001167  ENSG00000003056          2          173.096   \n",
       "1    ENSG00000001167  ENSG00000001629          2          166.671   \n",
       "2    ENSG00000001167  ENSG00000001561          2          132.995   \n",
       "3    ENSG00000001167  ENSG00000001460          2          125.436   \n",
       "4    ENSG00000001167  ENSG00000003393          2          131.500   \n",
       "..               ...              ...        ...              ...   \n",
       "106  ENSG00000002330  ENSG00000001036          2           11.414   \n",
       "107  ENSG00000001497  ENSG00000000460          2            8.448   \n",
       "108  ENSG00000002330  ENSG00000000460          2            7.225   \n",
       "109  ENSG00000001497  ENSG00000000005          2           17.364   \n",
       "110  ENSG00000001167  ENSG00000000005          2           10.929   \n",
       "\n",
       "     median_importance  \n",
       "0              178.614  \n",
       "1              168.557  \n",
       "2              137.291  \n",
       "3              128.146  \n",
       "4              141.749  \n",
       "..                 ...  \n",
       "106             12.249  \n",
       "107              8.633  \n",
       "108              8.388  \n",
       "109             26.507  \n",
       "110             17.483  \n",
       "\n",
       "[111 rows x 5 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "canonical_grn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "2bf35e68",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of edges in each category\n",
      "Number of edges in gene-exclusive:  25\n",
      "Number of edges in isoform-exclusive:  17577\n",
      "Number of edges in common interactions:  19\n"
     ]
    }
   ],
   "source": [
    "#filter aggregate\n",
    "as_aware_grn = filter_aggregated(as_aware_grn, threshold_frequency=1, threshold_importance=0.3)\n",
    "canonical_grn = filter_aggregated(canonical_grn, threshold_frequency=1, threshold_importance=0.3)\n",
    "\n",
    "net_AS = add_edge_key(as_aware_grn, biomart, source_column = 'source')\n",
    "net_canonical = add_edge_key(canonical_grn, biomart, type='canonical', source_column='source')\n",
    "\n",
    "# Categorize edges into gene-unique, isoform-unique, common\n",
    "common_edges = get_common_edges(net_canonical, net_AS)\n",
    "gene_unique, isoform_unique = get_diff(net_canonical, net_AS)\n",
    "\n",
    "print('Number of edges in each category')\n",
    "print('Number of edges in gene-exclusive: ', len(gene_unique))\n",
    "print('Number of edges in isoform-exclusive: ', len(isoform_unique))\n",
    "print('Number of edges in common interactions: ', len(common_edges))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "2f5a7301",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_transcript_data_source = common_edges.loc[:, ['source_gene', 'source_transcript']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e28f1273",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "adff8acf",
   "metadata": {},
   "outputs": [],
   "source": [
    "common_edges['source_descriptor'] = common_edges['source_gene']+':'+common_edges['source_transcript']\n",
    "common_edges['target_descriptor'] = common_edges['target_gene']+':'+common_edges['target_gene']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "d8557dd5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>frequency</th>\n",
       "      <th>mean_importance</th>\n",
       "      <th>median_importance</th>\n",
       "      <th>source_transcript</th>\n",
       "      <th>source_gene</th>\n",
       "      <th>source_transcript_name</th>\n",
       "      <th>source_gene_name</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>target_gene_name</th>\n",
       "      <th>type</th>\n",
       "      <th>edge_key</th>\n",
       "      <th>source_descriptor</th>\n",
       "      <th>target_descriptor</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000544271</td>\n",
       "      <td>ENSG00000002822</td>\n",
       "      <td>2</td>\n",
       "      <td>11.935</td>\n",
       "      <td>14.636</td>\n",
       "      <td>ENST00000544271</td>\n",
       "      <td>ENSG00000002330</td>\n",
       "      <td>BAD-206</td>\n",
       "      <td>BAD</td>\n",
       "      <td>ENSG00000002822</td>\n",
       "      <td>MAD1L1</td>\n",
       "      <td>AS</td>\n",
       "      <td>ENSG00000002330_ENSG00000002822</td>\n",
       "      <td>ENSG00000002330:ENST00000544271</td>\n",
       "      <td>ENSG00000002822:ENSG00000002822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000341376</td>\n",
       "      <td>ENSG00000001629</td>\n",
       "      <td>2</td>\n",
       "      <td>0.796</td>\n",
       "      <td>1.211</td>\n",
       "      <td>ENST00000341376</td>\n",
       "      <td>ENSG00000001167</td>\n",
       "      <td>NFYA-201</td>\n",
       "      <td>NFYA</td>\n",
       "      <td>ENSG00000001629</td>\n",
       "      <td>ANKIB1</td>\n",
       "      <td>AS</td>\n",
       "      <td>ENSG00000001167_ENSG00000001629</td>\n",
       "      <td>ENSG00000001167:ENST00000341376</td>\n",
       "      <td>ENSG00000001629:ENSG00000001629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000341376</td>\n",
       "      <td>ENSG00000003393</td>\n",
       "      <td>2</td>\n",
       "      <td>0.657</td>\n",
       "      <td>0.918</td>\n",
       "      <td>ENST00000341376</td>\n",
       "      <td>ENSG00000001167</td>\n",
       "      <td>NFYA-201</td>\n",
       "      <td>NFYA</td>\n",
       "      <td>ENSG00000003393</td>\n",
       "      <td>ALS2</td>\n",
       "      <td>AS</td>\n",
       "      <td>ENSG00000001167_ENSG00000003393</td>\n",
       "      <td>ENSG00000001167:ENST00000341376</td>\n",
       "      <td>ENSG00000003393:ENSG00000003393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000341376</td>\n",
       "      <td>ENSG00000003989</td>\n",
       "      <td>2</td>\n",
       "      <td>0.494</td>\n",
       "      <td>0.872</td>\n",
       "      <td>ENST00000341376</td>\n",
       "      <td>ENSG00000001167</td>\n",
       "      <td>NFYA-201</td>\n",
       "      <td>NFYA</td>\n",
       "      <td>ENSG00000003989</td>\n",
       "      <td>SLC7A2</td>\n",
       "      <td>AS</td>\n",
       "      <td>ENSG00000001167_ENSG00000003989</td>\n",
       "      <td>ENSG00000001167:ENST00000341376</td>\n",
       "      <td>ENSG00000003989:ENSG00000003989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000484069</td>\n",
       "      <td>ENSG00000003756</td>\n",
       "      <td>2</td>\n",
       "      <td>1.507</td>\n",
       "      <td>2.256</td>\n",
       "      <td>ENST00000484069</td>\n",
       "      <td>ENSG00000001497</td>\n",
       "      <td>LAS1L-205</td>\n",
       "      <td>LAS1L</td>\n",
       "      <td>ENSG00000003756</td>\n",
       "      <td>RBM5</td>\n",
       "      <td>AS</td>\n",
       "      <td>ENSG00000001497_ENSG00000003756</td>\n",
       "      <td>ENSG00000001497:ENST00000484069</td>\n",
       "      <td>ENSG00000003756:ENSG00000003756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000374811</td>\n",
       "      <td>ENSG00000003756</td>\n",
       "      <td>2</td>\n",
       "      <td>0.590</td>\n",
       "      <td>0.621</td>\n",
       "      <td>ENST00000374811</td>\n",
       "      <td>ENSG00000001497</td>\n",
       "      <td>LAS1L-203</td>\n",
       "      <td>LAS1L</td>\n",
       "      <td>ENSG00000003756</td>\n",
       "      <td>RBM5</td>\n",
       "      <td>AS</td>\n",
       "      <td>ENSG00000001497_ENSG00000003756</td>\n",
       "      <td>ENSG00000001497:ENST00000374811</td>\n",
       "      <td>ENSG00000003756:ENSG00000003756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000341376</td>\n",
       "      <td>ENSG00000003756</td>\n",
       "      <td>2</td>\n",
       "      <td>1.109</td>\n",
       "      <td>1.679</td>\n",
       "      <td>ENST00000341376</td>\n",
       "      <td>ENSG00000001167</td>\n",
       "      <td>NFYA-201</td>\n",
       "      <td>NFYA</td>\n",
       "      <td>ENSG00000003756</td>\n",
       "      <td>RBM5</td>\n",
       "      <td>AS</td>\n",
       "      <td>ENSG00000001167_ENSG00000003756</td>\n",
       "      <td>ENSG00000001167:ENST00000341376</td>\n",
       "      <td>ENSG00000003756:ENSG00000003756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000544271</td>\n",
       "      <td>ENSG00000003509</td>\n",
       "      <td>1</td>\n",
       "      <td>0.150</td>\n",
       "      <td>0.150</td>\n",
       "      <td>ENST00000544271</td>\n",
       "      <td>ENSG00000002330</td>\n",
       "      <td>BAD-206</td>\n",
       "      <td>BAD</td>\n",
       "      <td>ENSG00000003509</td>\n",
       "      <td>NDUFAF7</td>\n",
       "      <td>AS</td>\n",
       "      <td>ENSG00000002330_ENSG00000003509</td>\n",
       "      <td>ENSG00000002330:ENST00000544271</td>\n",
       "      <td>ENSG00000003509:ENSG00000003509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000341376</td>\n",
       "      <td>ENSG00000003056</td>\n",
       "      <td>2</td>\n",
       "      <td>1.418</td>\n",
       "      <td>2.807</td>\n",
       "      <td>ENST00000341376</td>\n",
       "      <td>ENSG00000001167</td>\n",
       "      <td>NFYA-201</td>\n",
       "      <td>NFYA</td>\n",
       "      <td>ENSG00000003056</td>\n",
       "      <td>M6PR</td>\n",
       "      <td>AS</td>\n",
       "      <td>ENSG00000001167_ENSG00000003056</td>\n",
       "      <td>ENSG00000001167:ENST00000341376</td>\n",
       "      <td>ENSG00000003056:ENSG00000003056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000341376</td>\n",
       "      <td>ENSG00000003402</td>\n",
       "      <td>1</td>\n",
       "      <td>0.254</td>\n",
       "      <td>0.254</td>\n",
       "      <td>ENST00000341376</td>\n",
       "      <td>ENSG00000001167</td>\n",
       "      <td>NFYA-201</td>\n",
       "      <td>NFYA</td>\n",
       "      <td>ENSG00000003402</td>\n",
       "      <td>CFLAR</td>\n",
       "      <td>AS</td>\n",
       "      <td>ENSG00000001167_ENSG00000003402</td>\n",
       "      <td>ENSG00000001167:ENST00000341376</td>\n",
       "      <td>ENSG00000003402:ENSG00000003402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENSG00000001167</td>\n",
       "      <td>ENSG00000003056</td>\n",
       "      <td>2</td>\n",
       "      <td>173.096</td>\n",
       "      <td>178.614</td>\n",
       "      <td>ENST00000341376</td>\n",
       "      <td>ENSG00000001167</td>\n",
       "      <td>NFYA-201</td>\n",
       "      <td>NFYA</td>\n",
       "      <td>ENSG00000003056</td>\n",
       "      <td>M6PR</td>\n",
       "      <td>canonical</td>\n",
       "      <td>ENSG00000001167_ENSG00000003056</td>\n",
       "      <td>ENSG00000001167:ENST00000341376</td>\n",
       "      <td>ENSG00000003056:ENSG00000003056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENSG00000001167</td>\n",
       "      <td>ENSG00000001629</td>\n",
       "      <td>2</td>\n",
       "      <td>166.671</td>\n",
       "      <td>168.557</td>\n",
       "      <td>ENST00000341376</td>\n",
       "      <td>ENSG00000001167</td>\n",
       "      <td>NFYA-201</td>\n",
       "      <td>NFYA</td>\n",
       "      <td>ENSG00000001629</td>\n",
       "      <td>ANKIB1</td>\n",
       "      <td>canonical</td>\n",
       "      <td>ENSG00000001167_ENSG00000001629</td>\n",
       "      <td>ENSG00000001167:ENST00000341376</td>\n",
       "      <td>ENSG00000001629:ENSG00000001629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENSG00000001167</td>\n",
       "      <td>ENSG00000003393</td>\n",
       "      <td>2</td>\n",
       "      <td>131.500</td>\n",
       "      <td>141.749</td>\n",
       "      <td>ENST00000341376</td>\n",
       "      <td>ENSG00000001167</td>\n",
       "      <td>NFYA-201</td>\n",
       "      <td>NFYA</td>\n",
       "      <td>ENSG00000003393</td>\n",
       "      <td>ALS2</td>\n",
       "      <td>canonical</td>\n",
       "      <td>ENSG00000001167_ENSG00000003393</td>\n",
       "      <td>ENSG00000001167:ENST00000341376</td>\n",
       "      <td>ENSG00000003393:ENSG00000003393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENSG00000002330</td>\n",
       "      <td>ENSG00000002822</td>\n",
       "      <td>2</td>\n",
       "      <td>110.318</td>\n",
       "      <td>113.245</td>\n",
       "      <td>ENST00000394532</td>\n",
       "      <td>ENSG00000002330</td>\n",
       "      <td>BAD-203</td>\n",
       "      <td>BAD</td>\n",
       "      <td>ENSG00000002822</td>\n",
       "      <td>MAD1L1</td>\n",
       "      <td>canonical</td>\n",
       "      <td>ENSG00000002330_ENSG00000002822</td>\n",
       "      <td>ENSG00000002330:ENST00000394532</td>\n",
       "      <td>ENSG00000002822:ENSG00000002822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENSG00000001167</td>\n",
       "      <td>ENSG00000003989</td>\n",
       "      <td>2</td>\n",
       "      <td>85.312</td>\n",
       "      <td>90.276</td>\n",
       "      <td>ENST00000341376</td>\n",
       "      <td>ENSG00000001167</td>\n",
       "      <td>NFYA-201</td>\n",
       "      <td>NFYA</td>\n",
       "      <td>ENSG00000003989</td>\n",
       "      <td>SLC7A2</td>\n",
       "      <td>canonical</td>\n",
       "      <td>ENSG00000001167_ENSG00000003989</td>\n",
       "      <td>ENSG00000001167:ENST00000341376</td>\n",
       "      <td>ENSG00000003989:ENSG00000003989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENSG00000001167</td>\n",
       "      <td>ENSG00000003756</td>\n",
       "      <td>2</td>\n",
       "      <td>94.662</td>\n",
       "      <td>108.511</td>\n",
       "      <td>ENST00000341376</td>\n",
       "      <td>ENSG00000001167</td>\n",
       "      <td>NFYA-201</td>\n",
       "      <td>NFYA</td>\n",
       "      <td>ENSG00000003756</td>\n",
       "      <td>RBM5</td>\n",
       "      <td>canonical</td>\n",
       "      <td>ENSG00000001167_ENSG00000003756</td>\n",
       "      <td>ENSG00000001167:ENST00000341376</td>\n",
       "      <td>ENSG00000003756:ENSG00000003756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENSG00000001497</td>\n",
       "      <td>ENSG00000003756</td>\n",
       "      <td>2</td>\n",
       "      <td>86.671</td>\n",
       "      <td>97.422</td>\n",
       "      <td>ENST00000677056</td>\n",
       "      <td>ENSG00000001497</td>\n",
       "      <td>LAS1L-207</td>\n",
       "      <td>LAS1L</td>\n",
       "      <td>ENSG00000003756</td>\n",
       "      <td>RBM5</td>\n",
       "      <td>canonical</td>\n",
       "      <td>ENSG00000001497_ENSG00000003756</td>\n",
       "      <td>ENSG00000001497:ENST00000677056</td>\n",
       "      <td>ENSG00000003756:ENSG00000003756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENSG00000001167</td>\n",
       "      <td>ENSG00000003402</td>\n",
       "      <td>2</td>\n",
       "      <td>70.120</td>\n",
       "      <td>76.850</td>\n",
       "      <td>ENST00000341376</td>\n",
       "      <td>ENSG00000001167</td>\n",
       "      <td>NFYA-201</td>\n",
       "      <td>NFYA</td>\n",
       "      <td>ENSG00000003402</td>\n",
       "      <td>CFLAR</td>\n",
       "      <td>canonical</td>\n",
       "      <td>ENSG00000001167_ENSG00000003402</td>\n",
       "      <td>ENSG00000001167:ENST00000341376</td>\n",
       "      <td>ENSG00000003402:ENSG00000003402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENSG00000002330</td>\n",
       "      <td>ENSG00000003509</td>\n",
       "      <td>2</td>\n",
       "      <td>53.370</td>\n",
       "      <td>57.937</td>\n",
       "      <td>ENST00000394532</td>\n",
       "      <td>ENSG00000002330</td>\n",
       "      <td>BAD-203</td>\n",
       "      <td>BAD</td>\n",
       "      <td>ENSG00000003509</td>\n",
       "      <td>NDUFAF7</td>\n",
       "      <td>canonical</td>\n",
       "      <td>ENSG00000002330_ENSG00000003509</td>\n",
       "      <td>ENSG00000002330:ENST00000394532</td>\n",
       "      <td>ENSG00000003509:ENSG00000003509</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             source           target  frequency  mean_importance  \\\n",
       "0   ENST00000544271  ENSG00000002822          2           11.935   \n",
       "1   ENST00000341376  ENSG00000001629          2            0.796   \n",
       "2   ENST00000341376  ENSG00000003393          2            0.657   \n",
       "3   ENST00000341376  ENSG00000003989          2            0.494   \n",
       "4   ENST00000484069  ENSG00000003756          2            1.507   \n",
       "5   ENST00000374811  ENSG00000003756          2            0.590   \n",
       "6   ENST00000341376  ENSG00000003756          2            1.109   \n",
       "7   ENST00000544271  ENSG00000003509          1            0.150   \n",
       "8   ENST00000341376  ENSG00000003056          2            1.418   \n",
       "9   ENST00000341376  ENSG00000003402          1            0.254   \n",
       "10  ENSG00000001167  ENSG00000003056          2          173.096   \n",
       "11  ENSG00000001167  ENSG00000001629          2          166.671   \n",
       "12  ENSG00000001167  ENSG00000003393          2          131.500   \n",
       "13  ENSG00000002330  ENSG00000002822          2          110.318   \n",
       "14  ENSG00000001167  ENSG00000003989          2           85.312   \n",
       "15  ENSG00000001167  ENSG00000003756          2           94.662   \n",
       "16  ENSG00000001497  ENSG00000003756          2           86.671   \n",
       "17  ENSG00000001167  ENSG00000003402          2           70.120   \n",
       "18  ENSG00000002330  ENSG00000003509          2           53.370   \n",
       "\n",
       "    median_importance source_transcript      source_gene  \\\n",
       "0              14.636   ENST00000544271  ENSG00000002330   \n",
       "1               1.211   ENST00000341376  ENSG00000001167   \n",
       "2               0.918   ENST00000341376  ENSG00000001167   \n",
       "3               0.872   ENST00000341376  ENSG00000001167   \n",
       "4               2.256   ENST00000484069  ENSG00000001497   \n",
       "5               0.621   ENST00000374811  ENSG00000001497   \n",
       "6               1.679   ENST00000341376  ENSG00000001167   \n",
       "7               0.150   ENST00000544271  ENSG00000002330   \n",
       "8               2.807   ENST00000341376  ENSG00000001167   \n",
       "9               0.254   ENST00000341376  ENSG00000001167   \n",
       "10            178.614   ENST00000341376  ENSG00000001167   \n",
       "11            168.557   ENST00000341376  ENSG00000001167   \n",
       "12            141.749   ENST00000341376  ENSG00000001167   \n",
       "13            113.245   ENST00000394532  ENSG00000002330   \n",
       "14             90.276   ENST00000341376  ENSG00000001167   \n",
       "15            108.511   ENST00000341376  ENSG00000001167   \n",
       "16             97.422   ENST00000677056  ENSG00000001497   \n",
       "17             76.850   ENST00000341376  ENSG00000001167   \n",
       "18             57.937   ENST00000394532  ENSG00000002330   \n",
       "\n",
       "   source_transcript_name source_gene_name      target_gene target_gene_name  \\\n",
       "0                 BAD-206              BAD  ENSG00000002822           MAD1L1   \n",
       "1                NFYA-201             NFYA  ENSG00000001629           ANKIB1   \n",
       "2                NFYA-201             NFYA  ENSG00000003393             ALS2   \n",
       "3                NFYA-201             NFYA  ENSG00000003989           SLC7A2   \n",
       "4               LAS1L-205            LAS1L  ENSG00000003756             RBM5   \n",
       "5               LAS1L-203            LAS1L  ENSG00000003756             RBM5   \n",
       "6                NFYA-201             NFYA  ENSG00000003756             RBM5   \n",
       "7                 BAD-206              BAD  ENSG00000003509          NDUFAF7   \n",
       "8                NFYA-201             NFYA  ENSG00000003056             M6PR   \n",
       "9                NFYA-201             NFYA  ENSG00000003402            CFLAR   \n",
       "10               NFYA-201             NFYA  ENSG00000003056             M6PR   \n",
       "11               NFYA-201             NFYA  ENSG00000001629           ANKIB1   \n",
       "12               NFYA-201             NFYA  ENSG00000003393             ALS2   \n",
       "13                BAD-203              BAD  ENSG00000002822           MAD1L1   \n",
       "14               NFYA-201             NFYA  ENSG00000003989           SLC7A2   \n",
       "15               NFYA-201             NFYA  ENSG00000003756             RBM5   \n",
       "16              LAS1L-207            LAS1L  ENSG00000003756             RBM5   \n",
       "17               NFYA-201             NFYA  ENSG00000003402            CFLAR   \n",
       "18                BAD-203              BAD  ENSG00000003509          NDUFAF7   \n",
       "\n",
       "         type                         edge_key  \\\n",
       "0          AS  ENSG00000002330_ENSG00000002822   \n",
       "1          AS  ENSG00000001167_ENSG00000001629   \n",
       "2          AS  ENSG00000001167_ENSG00000003393   \n",
       "3          AS  ENSG00000001167_ENSG00000003989   \n",
       "4          AS  ENSG00000001497_ENSG00000003756   \n",
       "5          AS  ENSG00000001497_ENSG00000003756   \n",
       "6          AS  ENSG00000001167_ENSG00000003756   \n",
       "7          AS  ENSG00000002330_ENSG00000003509   \n",
       "8          AS  ENSG00000001167_ENSG00000003056   \n",
       "9          AS  ENSG00000001167_ENSG00000003402   \n",
       "10  canonical  ENSG00000001167_ENSG00000003056   \n",
       "11  canonical  ENSG00000001167_ENSG00000001629   \n",
       "12  canonical  ENSG00000001167_ENSG00000003393   \n",
       "13  canonical  ENSG00000002330_ENSG00000002822   \n",
       "14  canonical  ENSG00000001167_ENSG00000003989   \n",
       "15  canonical  ENSG00000001167_ENSG00000003756   \n",
       "16  canonical  ENSG00000001497_ENSG00000003756   \n",
       "17  canonical  ENSG00000001167_ENSG00000003402   \n",
       "18  canonical  ENSG00000002330_ENSG00000003509   \n",
       "\n",
       "                  source_descriptor                target_descriptor  \n",
       "0   ENSG00000002330:ENST00000544271  ENSG00000002822:ENSG00000002822  \n",
       "1   ENSG00000001167:ENST00000341376  ENSG00000001629:ENSG00000001629  \n",
       "2   ENSG00000001167:ENST00000341376  ENSG00000003393:ENSG00000003393  \n",
       "3   ENSG00000001167:ENST00000341376  ENSG00000003989:ENSG00000003989  \n",
       "4   ENSG00000001497:ENST00000484069  ENSG00000003756:ENSG00000003756  \n",
       "5   ENSG00000001497:ENST00000374811  ENSG00000003756:ENSG00000003756  \n",
       "6   ENSG00000001167:ENST00000341376  ENSG00000003756:ENSG00000003756  \n",
       "7   ENSG00000002330:ENST00000544271  ENSG00000003509:ENSG00000003509  \n",
       "8   ENSG00000001167:ENST00000341376  ENSG00000003056:ENSG00000003056  \n",
       "9   ENSG00000001167:ENST00000341376  ENSG00000003402:ENSG00000003402  \n",
       "10  ENSG00000001167:ENST00000341376  ENSG00000003056:ENSG00000003056  \n",
       "11  ENSG00000001167:ENST00000341376  ENSG00000001629:ENSG00000001629  \n",
       "12  ENSG00000001167:ENST00000341376  ENSG00000003393:ENSG00000003393  \n",
       "13  ENSG00000002330:ENST00000394532  ENSG00000002822:ENSG00000002822  \n",
       "14  ENSG00000001167:ENST00000341376  ENSG00000003989:ENSG00000003989  \n",
       "15  ENSG00000001167:ENST00000341376  ENSG00000003756:ENSG00000003756  \n",
       "16  ENSG00000001497:ENST00000677056  ENSG00000003756:ENSG00000003756  \n",
       "17  ENSG00000001167:ENST00000341376  ENSG00000003402:ENSG00000003402  \n",
       "18  ENSG00000002330:ENST00000394532  ENSG00000003509:ENSG00000003509  "
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "common_edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d5f18b4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENSG00000002330\n",
      "ENST00000544271\n",
      "ENSG00000001167\n",
      "ENST00000341376\n",
      "ENSG00000001167\n",
      "ENST00000341376\n",
      "ENSG00000001167\n",
      "ENST00000341376\n",
      "ENSG00000001497\n",
      "ENST00000484069\n",
      "ENSG00000001497\n",
      "ENST00000374811\n",
      "ENSG00000001167\n",
      "ENST00000341376\n",
      "ENSG00000002330\n",
      "ENST00000544271\n",
      "ENSG00000001167\n",
      "ENST00000341376\n",
      "ENSG00000001167\n",
      "ENST00000341376\n",
      "ENSG00000001167\n",
      "ENST00000341376\n",
      "ENSG00000001167\n",
      "ENST00000341376\n",
      "ENSG00000001167\n",
      "ENST00000341376\n",
      "ENSG00000002330\n",
      "ENST00000394532\n",
      "ENSG00000001167\n",
      "ENST00000341376\n",
      "ENSG00000001167\n",
      "ENST00000341376\n",
      "ENSG00000001497\n",
      "ENST00000677056\n",
      "ENSG00000001167\n",
      "ENST00000341376\n",
      "ENSG00000002330\n",
      "ENST00000394532\n"
     ]
    }
   ],
   "source": [
    "gene_transcript_map = defaultdict(set)\n",
    "\n",
    "# Iterate through each pair of (gene, transcript) from the DataFrame using zip.\n",
    "for gene, transcript in zip(gene_transcript_data_source['source_gene'], gene_transcript_data_source['source_transcript']):\n",
    "    # Add the transcript to the set associated with the current gene.\n",
    "    # The set automatically handles duplicate transcripts, ensuring each\n",
    "    # transcript is listed only once for a given gene.\n",
    "    gene_transcript_map[gene].add(transcript)\n",
    "\n",
    "\n",
    "# Convert the sets of transcripts into sorted lists to match the desired output format.\n",
    "# A dictionary comprehension is used for a clean and concise conversion.\n",
    "node_data = {\n",
    "    gene: sorted(list(transcripts))\n",
    "    for gene, transcripts in gene_transcript_map.items()\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "a329d555",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "defaultdict(set,\n",
       "            {'ENSG00000002330': {'ENST00000394532', 'ENST00000544271'},\n",
       "             'ENSG00000001167': {'ENST00000341376'},\n",
       "             'ENSG00000001497': {'ENST00000374811',\n",
       "              'ENST00000484069',\n",
       "              'ENST00000677056'}})"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_transcript_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7fbc58c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: node ENSG00000002822, port ENSG00000002822 unrecognized\n",
      "Warning: node ENSG00000002822, port ENSG00000002822 unrecognized\n",
      "Warning: node ENSG00000003509, port ENSG00000003509 unrecognized\n",
      "Warning: node ENSG00000003509, port ENSG00000003509 unrecognized\n",
      "Warning: node ENSG00000001629, port ENSG00000001629 unrecognized\n",
      "Warning: node ENSG00000001629, port ENSG00000001629 unrecognized\n",
      "Warning: node ENSG00000003393, port ENSG00000003393 unrecognized\n",
      "Warning: node ENSG00000003393, port ENSG00000003393 unrecognized\n",
      "Warning: node ENSG00000003989, port ENSG00000003989 unrecognized\n",
      "Warning: node ENSG00000003989, port ENSG00000003989 unrecognized\n",
      "Warning: node ENSG00000003756, port ENSG00000003756 unrecognized\n",
      "Warning: node ENSG00000003756, port ENSG00000003756 unrecognized\n",
      "Warning: node ENSG00000003056, port ENSG00000003056 unrecognized\n",
      "Warning: node ENSG00000003056, port ENSG00000003056 unrecognized\n",
      "Warning: node ENSG00000003402, port ENSG00000003402 unrecognized\n",
      "Warning: node ENSG00000003402, port ENSG00000003402 unrecognized\n",
      "Warning: node ENSG00000003756, port ENSG00000003756 unrecognized\n",
      "Warning: node ENSG00000003756, port ENSG00000003756 unrecognized\n",
      "Warning: node ENSG00000003756, port ENSG00000003756 unrecognized\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'gene_transcripts_expanded.gv.pdf'"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Error: no \"view\" rule for type \"application/pdf\" passed its test case\n",
      "       (for more information, add \"--debug=1\" on the command line)\n"
     ]
    }
   ],
   "source": [
    "import graphviz\n",
    "\n",
    "s = graphviz.Digraph(\n",
    "    'gene_transcripts_expanded',\n",
    "    filename='gene_transcripts_expanded.gv',\n",
    "    graph_attr={\n",
    "        'splines': 'spline',  # Added splines attribute\n",
    "        'rankdir': 'neato'\n",
    "    }\n",
    ")\n",
    "# Set global node and edge attributes for a consistent look.\n",
    "s.attr('node', shape='record', style='rounded, filled', height='0.1')\n",
    "s.attr('edge', color='grey', arrowhead='vee')\n",
    "\n",
    "# Assume common_edges is your data source\n",
    "# For this example, let's use a simple dictionary\n",
    "\n",
    "for gene, transcripts in node_data.items():\n",
    "    # Create a list of formatted transcript strings with ports\n",
    "    transcript_ports = [f\"<{t}>{t}\" for t in transcripts]\n",
    "\n",
    "    # Join the gene and the formatted transcripts for the node label\n",
    "    label = f\"{{<{gene}>{gene}|{'|'.join(transcript_ports)}}}\"\n",
    "\n",
    "    # Use the new formatted string to create the node\n",
    "    s.node(gene, label)\n",
    "\n",
    "for idx, row in common_edges.iterrows():\n",
    "\n",
    "    # Add some edges to demonstrate the 'ortho' splines\n",
    "    s.edge(row['source_descriptor'], row['target_descriptor'])\n",
    "\n",
    "# Render the graph\n",
    "s.render(view=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4588257",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'gene_transcripts_expanded.gv.pdf'"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Error: no \"view\" rule for type \"application/pdf\" passed its test case\n",
      "       (for more information, add \"--debug=1\" on the command line)\n"
     ]
    }
   ],
   "source": [
    "s.view()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b52bc78a",
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'graphviz'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mModuleNotFoundError\u001b[39m                       Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[23]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mgraphviz\u001b[39;00m\n\u001b[32m      3\u001b[39m \u001b[38;5;66;03m# Create a new directed graph with an increased number of nodes.\u001b[39;00m\n\u001b[32m      4\u001b[39m \u001b[38;5;66;03m# The 'splines' attribute with 'ortho' makes the edges straight.\u001b[39;00m\n\u001b[32m      5\u001b[39m s = graphviz.Digraph(\n\u001b[32m      6\u001b[39m     \u001b[33m'\u001b[39m\u001b[33mgene_transcripts_expanded\u001b[39m\u001b[33m'\u001b[39m,\n\u001b[32m      7\u001b[39m     filename=\u001b[33m'\u001b[39m\u001b[33mgene_transcripts_expanded.gv\u001b[39m\u001b[33m'\u001b[39m,\n\u001b[32m      8\u001b[39m     graph_attr={\u001b[33m'\u001b[39m\u001b[33msplines\u001b[39m\u001b[33m'\u001b[39m: \u001b[33m'\u001b[39m\u001b[33mortho\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mrankdir\u001b[39m\u001b[33m'\u001b[39m: \u001b[33m'\u001b[39m\u001b[33mneato\u001b[39m\u001b[33m'\u001b[39m}\n\u001b[32m      9\u001b[39m )\n",
      "\u001b[31mModuleNotFoundError\u001b[39m: No module named 'graphviz'"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "# Create a new directed graph with an increased number of nodes.\n",
    "# The 'splines' attribute with 'ortho' makes the edges straight.\n",
    "s = graphviz.Digraph(\n",
    "    'gene_transcripts_expanded',\n",
    "    filename='gene_transcripts_expanded.gv',\n",
    "    graph_attr={'splines': 'ortho', 'rankdir': 'neato'}\n",
    ")\n",
    "\n",
    "# Set global node and edge attributes for a consistent look.\n",
    "s.attr('node', shape='record', style='rounded, filled', height='0.1')\n",
    "s.attr('edge', color='grey', arrowhead='vee')\n",
    "s.node('A_T3', 'Transcript MYC-003 | <here> variant | coding | intron')\n",
    "s.view()\n",
    "# --- Define the nodes for the genes and their transcripts (20 nodes total) ---\n",
    "\n",
    "# Define the 'Gene_A' node. This is the top-level entity.\n",
    "s.node('Gene_A', 'Gene MYC', style='filled', fillcolor='#c5e1a5')\n",
    "\n",
    "# Define the 'Gene_B' node. Another top-level gene.\n",
    "s.node('Gene_B', 'Gene TP53', style='filled', fillcolor='#c5e1a5')\n",
    "\n",
    "# Define the 'Gene_C' node. A third gene.\n",
    "s.node('Gene_C', 'Gene BCL2', style='filled', fillcolor='#c5e1a5')\n",
    "\n",
    "# Define the 'Gene_D' node. A fourth gene.\n",
    "s.node('Gene_D', 'Gene STAT3', style='filled', fillcolor='#c5e1a5')\n",
    "\n",
    "# Define transcripts for Gene_A (4 transcripts).\n",
    "s.node('A_T1', 'Transcript MYC-001')\n",
    "s.node('A_T2', 'Transcript MYC-002')\n",
    "s.node('A_T3', 'Transcript MYC-003 | <here> variant | coding | intron')\n",
    "s.node('A_T4', 'Transcript MYC-004')\n",
    "\n",
    "# Define transcripts for Gene_B (5 transcripts).\n",
    "s.node('B_T1', 'Transcript TP53-001')\n",
    "s.node('B_T2', 'Transcript TP53-002')\n",
    "s.node('B_T3', 'Transcript TP53-003')\n",
    "s.node('B_T4', 'Transcript TP53-004')\n",
    "s.node('B_T5', 'Transcript TP53-005')\n",
    "\n",
    "# Define transcripts for Gene_C (4 transcripts).\n",
    "s.node('C_T1', 'Transcript BCL2-001')\n",
    "s.node('C_T2', 'Transcript BCL2-002')\n",
    "s.node('C_T3', 'Transcript BCL2-003')\n",
    "s.node('C_T4', 'Transcript BCL2-004')\n",
    "\n",
    "# Define transcripts for Gene_D (4 transcripts).\n",
    "s.node('D_T1', 'Transcript STAT3-001')\n",
    "s.node('D_T2', 'Transcript STAT3-002')\n",
    "s.node('D_T3', 'Transcript STAT3-003')\n",
    "s.node('D_T4', 'Transcript STAT3-004')\n",
    "\n",
    "# --- Define the edges to show the relationships ---\n",
    "\n",
    "# Edges from Gene_A to its transcripts.\n",
    "s.edge('Gene_A', 'A_T1')\n",
    "s.edge('Gene_A', 'A_T2')\n",
    "s.edge('Gene_A', 'A_T3')\n",
    "s.edge('Gene_A', 'A_T4')\n",
    "\n",
    "# Edges from Gene_B to its transcripts.\n",
    "s.edge('Gene_B', 'B_T1')\n",
    "s.edge('Gene_B', 'B_T2')\n",
    "s.edge('Gene_B', 'B_T3')\n",
    "s.edge('Gene_B', 'B_T4')\n",
    "s.edge('Gene_B', 'B_T5')\n",
    "\n",
    "# Edges from Gene_C to its transcripts.\n",
    "s.edge('Gene_C', 'C_T1')\n",
    "s.edge('Gene_C', 'C_T2')\n",
    "s.edge('Gene_C', 'C_T3')\n",
    "s.edge('Gene_C', 'C_T4')\n",
    "\n",
    "# Edges from Gene_D to its transcripts.\n",
    "s.edge('Gene_D', 'D_T1')\n",
    "s.edge('Gene_D', 'D_T2')\n",
    "s.edge('Gene_D', 'D_T3')\n",
    "s.edge('Gene_D', 'D_T4')\n",
    "\n",
    "# Add specific relationships like alternative splicing and regulatory interactions.\n",
    "s.edge('A_T2', 'A_T3:here', label='alternative splicing')\n",
    "s.edge('B_T1', 'C_T2', label='regulatory interaction')\n",
    "s.edge('C_T3', 'D_T1', label='regulatory interaction')\n",
    "s.edge('D_T2', 'A_T4', label='regulatory interaction')\n",
    "\n",
    "# Render the graph to a file and open it for viewing.\n",
    "s.view()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7d7e081",
   "metadata": {},
   "outputs": [],
   "source": [
    "plausibility_filtered = plausibility_filtering(config, isoform_unique, isoform_categories, tf_database)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79f20181",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pixi_alternet",
   "language": "python",
   "name": "pixi_alternet"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
