{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "fc5c48b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import pandas as pd\n",
    "from load_data import map_tf_ids, separate_tf_genes, prepare_for_inference\n",
    "import db as db\n",
    "from utils_network import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "056be149",
   "metadata": {},
   "outputs": [],
   "source": [
    "config_file = \"/data_nfs/og86asub/alternet-project/alternet/configs/MAGNet_NF.yaml\"\n",
    "import yaml\n",
    "\n",
    "with open(config_file, 'r') as f:\n",
    "    config = yaml.safe_load(f)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f120a43d",
   "metadata": {},
   "source": [
    "# Load necessary data and annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "4e63cb94",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data_nfs/og86asub/alternet-project/alternet/src/alternet/db.py:32: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  digger = pd.read_csv(digger_path)\n"
     ]
    }
   ],
   "source": [
    "biomart_path = '/data_nfs/og86asub/alternet-project/alternet/data/biomart.txt'\n",
    "tf_list_path = '/data_nfs/og86asub/alternet-project/alternet/data/allTFs_hg38.txt'\n",
    "\n",
    "biomart = pd.read_csv(biomart_path, sep='\\t')\n",
    "tf_list = pd.read_csv(tf_list_path, sep='\\t', header = None)\n",
    "tf_list = map_tf_ids(tf_list, biomart)\n",
    "\n",
    "# create TF Database\n",
    "tf_database = db.create_transcipt_annotation_database(tf_list=tf_list, appris_path= config['appris'], digger_path=config['digger'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "6fd4d8d6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TF</th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Transcript stable ID</th>\n",
       "      <th>Gene name (HGNC)</th>\n",
       "      <th>Translation ID</th>\n",
       "      <th>Is translated?</th>\n",
       "      <th>Transcript type</th>\n",
       "      <th>Not found tag</th>\n",
       "      <th>CCDS ID</th>\n",
       "      <th>Transcript support level</th>\n",
       "      <th>...</th>\n",
       "      <th>Trans-membrane helices (THUMP)</th>\n",
       "      <th>Signal sequence (CRASH)</th>\n",
       "      <th>Trifid Score</th>\n",
       "      <th>Peptides</th>\n",
       "      <th>APPRIS score</th>\n",
       "      <th>APPRIS Annotation</th>\n",
       "      <th>MANE</th>\n",
       "      <th>Exon stable ID</th>\n",
       "      <th>Exon rank in transcript</th>\n",
       "      <th>Pfam ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "      <td>ENST00000373997</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENSP00000363109</td>\n",
       "      <td>TRANSLATION</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>-</td>\n",
       "      <td>CCDS7241.1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-4,-5</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>29</td>\n",
       "      <td>12.000</td>\n",
       "      <td>ALTERNATIVE:1</td>\n",
       "      <td>MANE_Select</td>\n",
       "      <td>[ENSE00003539976, ENSE00003688232, ENSE0000368...</td>\n",
       "      <td>[3, 4, 4, 5, 5, 6, 6, 6, 7, 7, 8, 8, 9, 9, 10,...</td>\n",
       "      <td>[PF00076, PF00076, PF00076, PF00076, PF00076, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "      <td>ENST00000374001</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENSP00000363113</td>\n",
       "      <td>TRANSLATION</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>-</td>\n",
       "      <td>CCDS7241.1</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-4,-5</td>\n",
       "      <td>0.9986</td>\n",
       "      <td>29</td>\n",
       "      <td>12.000</td>\n",
       "      <td>ALTERNATIVE:1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENSE00001345632, ENSE00003539976, ENSE0000368...</td>\n",
       "      <td>[1, 2, 3, 3, 4, 4, 5, 5, 5, 6, 6, 7, 7, 8, 8, ...</td>\n",
       "      <td>[PF00076, PF00076, PF00076, PF00076, PF00076, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "      <td>ENST00000414883</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENSP00000397953</td>\n",
       "      <td>TRANSLATION</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>-</td>\n",
       "      <td>CCDS7241.1</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-4,-5</td>\n",
       "      <td>0.9373</td>\n",
       "      <td>29</td>\n",
       "      <td>12.000</td>\n",
       "      <td>ALTERNATIVE:1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENSE00001345632, ENSE00003539976, ENSE0000368...</td>\n",
       "      <td>[1, 4, 5, 5, 6, 6, 7, 7, 2, 3]</td>\n",
       "      <td>[PF00076, PF00076, PF00076, PF00076]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "      <td>ENST00000373993</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENSP00000363105</td>\n",
       "      <td>TRANSLATION</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>-</td>\n",
       "      <td>CCDS7242.1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-4,-5</td>\n",
       "      <td>0.9733</td>\n",
       "      <td>30</td>\n",
       "      <td>11.508</td>\n",
       "      <td>PRINCIPAL:3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENSE00003539976, ENSE00003688232, ENSE0000368...</td>\n",
       "      <td>[1, 2, 2, 3, 3, 4, 4, 4, 5, 5, 6, 6, 7, 7, 9, ...</td>\n",
       "      <td>[PF00076, PF00076, PF00076, PF00076, PF00076, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "      <td>ENST00000373995</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENSP00000363107</td>\n",
       "      <td>TRANSLATION</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>-</td>\n",
       "      <td>CCDS7243.1</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-4,-5</td>\n",
       "      <td>0.2064</td>\n",
       "      <td>27</td>\n",
       "      <td>6.998</td>\n",
       "      <td>MINOR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENSE00001345632, ENSE00003688232, ENSE0000368...</td>\n",
       "      <td>[1, 4, 4, 5, 5, 6, 6, 6, 7, 7, 8, 8, 9, 9, 10,...</td>\n",
       "      <td>[PF00076, PF00076, PF00076, PF00076, PF00076, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16293</th>\n",
       "      <td>ZXDB</td>\n",
       "      <td>ENSG00000198455</td>\n",
       "      <td>ENST00000374888</td>\n",
       "      <td>ZXDB</td>\n",
       "      <td>ENSP00000364023</td>\n",
       "      <td>TRANSLATION</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>-</td>\n",
       "      <td>CCDS35313.1</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-4,-5</td>\n",
       "      <td>0.8479</td>\n",
       "      <td>-</td>\n",
       "      <td>13.000</td>\n",
       "      <td>PRINCIPAL:1</td>\n",
       "      <td>MANE_Select</td>\n",
       "      <td>[ENSE00001465020, ENSE00001465020, ENSE0000146...</td>\n",
       "      <td>[1, 1, 1, 1, 1]</td>\n",
       "      <td>[PF00096, PF00096, PF00096, PF00096, PF00096]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16294</th>\n",
       "      <td>ZXDC</td>\n",
       "      <td>ENSG00000070476</td>\n",
       "      <td>ENST00000514463</td>\n",
       "      <td>ZXDC</td>\n",
       "      <td>-</td>\n",
       "      <td>NO_TRANSLATION</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16295</th>\n",
       "      <td>ZXDC</td>\n",
       "      <td>ENSG00000070476</td>\n",
       "      <td>ENST00000389709</td>\n",
       "      <td>ZXDC</td>\n",
       "      <td>ENSP00000374359</td>\n",
       "      <td>TRANSLATION</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>-</td>\n",
       "      <td>CCDS43145.1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-4,-5</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>1</td>\n",
       "      <td>15.000</td>\n",
       "      <td>PRINCIPAL:2</td>\n",
       "      <td>MANE_Select</td>\n",
       "      <td>[ENSE00001506699, ENSE00001506699, ENSE0000150...</td>\n",
       "      <td>[1, 1, 1, 2, 3, 4, 4, 5, 6, 7, 8, 9, 10]</td>\n",
       "      <td>[PF00096, PF00096, PF00096]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16296</th>\n",
       "      <td>ZXDC</td>\n",
       "      <td>ENSG00000070476</td>\n",
       "      <td>ENST00000515545</td>\n",
       "      <td>ZXDC</td>\n",
       "      <td>ENSP00000426532</td>\n",
       "      <td>TRANSLATION</td>\n",
       "      <td>nonsense_mediated_decay</td>\n",
       "      <td>start</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-4,-5</td>\n",
       "      <td>0.0398</td>\n",
       "      <td>1</td>\n",
       "      <td>-2.000</td>\n",
       "      <td>MINOR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16297</th>\n",
       "      <td>ZXDC</td>\n",
       "      <td>ENSG00000070476</td>\n",
       "      <td>ENST00000336332</td>\n",
       "      <td>ZXDC</td>\n",
       "      <td>ENSP00000337694</td>\n",
       "      <td>TRANSLATION</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>-</td>\n",
       "      <td>CCDS43146.1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-4,-5</td>\n",
       "      <td>0.7370</td>\n",
       "      <td>1</td>\n",
       "      <td>14.000</td>\n",
       "      <td>ALTERNATIVE:2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENSE00001506698, ENSE00001506697, ENSE0000150...</td>\n",
       "      <td>[2, 3, 3, 4, 4, 4, 5, 1, 1, 1, 6]</td>\n",
       "      <td>[PF00096, PF00096, PF00096, PF00096, PF00096]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16298 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         TF   Gene stable ID Transcript stable ID Gene name (HGNC)  \\\n",
       "0      A1CF  ENSG00000148584      ENST00000373997             A1CF   \n",
       "1      A1CF  ENSG00000148584      ENST00000374001             A1CF   \n",
       "2      A1CF  ENSG00000148584      ENST00000414883             A1CF   \n",
       "3      A1CF  ENSG00000148584      ENST00000373993             A1CF   \n",
       "4      A1CF  ENSG00000148584      ENST00000373995             A1CF   \n",
       "...     ...              ...                  ...              ...   \n",
       "16293  ZXDB  ENSG00000198455      ENST00000374888             ZXDB   \n",
       "16294  ZXDC  ENSG00000070476      ENST00000514463             ZXDC   \n",
       "16295  ZXDC  ENSG00000070476      ENST00000389709             ZXDC   \n",
       "16296  ZXDC  ENSG00000070476      ENST00000515545             ZXDC   \n",
       "16297  ZXDC  ENSG00000070476      ENST00000336332             ZXDC   \n",
       "\n",
       "        Translation ID  Is translated?          Transcript type Not found tag  \\\n",
       "0      ENSP00000363109     TRANSLATION           protein_coding             -   \n",
       "1      ENSP00000363113     TRANSLATION           protein_coding             -   \n",
       "2      ENSP00000397953     TRANSLATION           protein_coding             -   \n",
       "3      ENSP00000363105     TRANSLATION           protein_coding             -   \n",
       "4      ENSP00000363107     TRANSLATION           protein_coding             -   \n",
       "...                ...             ...                      ...           ...   \n",
       "16293  ENSP00000364023     TRANSLATION           protein_coding             -   \n",
       "16294                -  NO_TRANSLATION          retained_intron           NaN   \n",
       "16295  ENSP00000374359     TRANSLATION           protein_coding             -   \n",
       "16296  ENSP00000426532     TRANSLATION  nonsense_mediated_decay         start   \n",
       "16297  ENSP00000337694     TRANSLATION           protein_coding             -   \n",
       "\n",
       "           CCDS ID Transcript support level  ...  \\\n",
       "0       CCDS7241.1                        1  ...   \n",
       "1       CCDS7241.1                        5  ...   \n",
       "2       CCDS7241.1                        5  ...   \n",
       "3       CCDS7242.1                        1  ...   \n",
       "4       CCDS7243.1                        2  ...   \n",
       "...            ...                      ...  ...   \n",
       "16293  CCDS35313.1                        -  ...   \n",
       "16294          NaN                      NaN  ...   \n",
       "16295  CCDS43145.1                        1  ...   \n",
       "16296            -                        1  ...   \n",
       "16297  CCDS43146.1                        1  ...   \n",
       "\n",
       "       Trans-membrane helices (THUMP)  Signal sequence (CRASH)  Trifid Score  \\\n",
       "0                                 0.0                    -4,-5        1.0000   \n",
       "1                                 0.0                    -4,-5        0.9986   \n",
       "2                                 0.0                    -4,-5        0.9373   \n",
       "3                                 0.0                    -4,-5        0.9733   \n",
       "4                                 0.0                    -4,-5        0.2064   \n",
       "...                               ...                      ...           ...   \n",
       "16293                             0.0                    -4,-5        0.8479   \n",
       "16294                             NaN                      NaN           NaN   \n",
       "16295                             0.0                    -4,-5        1.0000   \n",
       "16296                             0.0                    -4,-5        0.0398   \n",
       "16297                             0.0                    -4,-5        0.7370   \n",
       "\n",
       "       Peptides  APPRIS score  APPRIS Annotation         MANE  \\\n",
       "0            29        12.000      ALTERNATIVE:1  MANE_Select   \n",
       "1            29        12.000      ALTERNATIVE:1          NaN   \n",
       "2            29        12.000      ALTERNATIVE:1          NaN   \n",
       "3            30        11.508        PRINCIPAL:3          NaN   \n",
       "4            27         6.998              MINOR          NaN   \n",
       "...         ...           ...                ...          ...   \n",
       "16293         -        13.000        PRINCIPAL:1  MANE_Select   \n",
       "16294       NaN           NaN                NaN          NaN   \n",
       "16295         1        15.000        PRINCIPAL:2  MANE_Select   \n",
       "16296         1        -2.000              MINOR          NaN   \n",
       "16297         1        14.000      ALTERNATIVE:2          NaN   \n",
       "\n",
       "                                          Exon stable ID  \\\n",
       "0      [ENSE00003539976, ENSE00003688232, ENSE0000368...   \n",
       "1      [ENSE00001345632, ENSE00003539976, ENSE0000368...   \n",
       "2      [ENSE00001345632, ENSE00003539976, ENSE0000368...   \n",
       "3      [ENSE00003539976, ENSE00003688232, ENSE0000368...   \n",
       "4      [ENSE00001345632, ENSE00003688232, ENSE0000368...   \n",
       "...                                                  ...   \n",
       "16293  [ENSE00001465020, ENSE00001465020, ENSE0000146...   \n",
       "16294                                                NaN   \n",
       "16295  [ENSE00001506699, ENSE00001506699, ENSE0000150...   \n",
       "16296                                                NaN   \n",
       "16297  [ENSE00001506698, ENSE00001506697, ENSE0000150...   \n",
       "\n",
       "                                 Exon rank in transcript  \\\n",
       "0      [3, 4, 4, 5, 5, 6, 6, 6, 7, 7, 8, 8, 9, 9, 10,...   \n",
       "1      [1, 2, 3, 3, 4, 4, 5, 5, 5, 6, 6, 7, 7, 8, 8, ...   \n",
       "2                         [1, 4, 5, 5, 6, 6, 7, 7, 2, 3]   \n",
       "3      [1, 2, 2, 3, 3, 4, 4, 4, 5, 5, 6, 6, 7, 7, 9, ...   \n",
       "4      [1, 4, 4, 5, 5, 6, 6, 6, 7, 7, 8, 8, 9, 9, 10,...   \n",
       "...                                                  ...   \n",
       "16293                                    [1, 1, 1, 1, 1]   \n",
       "16294                                                NaN   \n",
       "16295           [1, 1, 1, 2, 3, 4, 4, 5, 6, 7, 8, 9, 10]   \n",
       "16296                                                NaN   \n",
       "16297                  [2, 3, 3, 4, 4, 4, 5, 1, 1, 1, 6]   \n",
       "\n",
       "                                                 Pfam ID  \n",
       "0      [PF00076, PF00076, PF00076, PF00076, PF00076, ...  \n",
       "1      [PF00076, PF00076, PF00076, PF00076, PF00076, ...  \n",
       "2                   [PF00076, PF00076, PF00076, PF00076]  \n",
       "3      [PF00076, PF00076, PF00076, PF00076, PF00076, ...  \n",
       "4      [PF00076, PF00076, PF00076, PF00076, PF00076, ...  \n",
       "...                                                  ...  \n",
       "16293      [PF00096, PF00096, PF00096, PF00096, PF00096]  \n",
       "16294                                                NaN  \n",
       "16295                        [PF00096, PF00096, PF00096]  \n",
       "16296                                                NaN  \n",
       "16297      [PF00096, PF00096, PF00096, PF00096, PF00096]  \n",
       "\n",
       "[16298 rows x 25 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tf_database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "3125cafc",
   "metadata": {},
   "outputs": [],
   "source": [
    "## LOAD gene and transcript data\n",
    "transcript_data = pd.read_csv(config['transcript_data'], index_col=0)\n",
    "\n",
    "## GET the gene ids associated with transcripts\n",
    "gene_ids = transcript_data['gene_id'].unique()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "8b2429d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import rnanorm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "6f71ada4",
   "metadata": {},
   "outputs": [],
   "source": [
    "## REad the gene data\n",
    "gene_data = pd.read_csv(config['count_data'], index_col=0)\n",
    "gene_data.index.name = 'gene_id'\n",
    "gene_data = gene_data.reset_index()\n",
    "gene_data = gene_data[gene_data['gene_id'].isin(gene_ids)]\n",
    "\n",
    "gene_data = gene_data.set_index('gene_id')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "e2928dcb",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_data = gene_data.iloc[0:20:, ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "e6790020",
   "metadata": {},
   "outputs": [],
   "source": [
    "fpkm = rnanorm.FPKM('/data_nfs/datasets/GRCh38/Homo_sapiens.GRCh38.109.gtf').set_output(transform=\"pandas\")\n",
    "gene_data = fpkm.fit_transform(gene_data.T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "15ca18c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_data= gene_data.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "70b0d1ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_data.index.name = 'gene_id'\n",
    "gene_data = gene_data.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "902ba019",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "## Subset the samples of interest\n",
    "sample_attributes = pd.read_csv(config['sample_attributes'])\n",
    "sample_attributes = sample_attributes.loc[:, ['sample_name', 'etiology']]\n",
    "samples = sample_attributes[sample_attributes['etiology'] == config['tissue']]\n",
    "samples = samples['sample_name'].tolist()\n",
    "\n",
    "\n",
    "## Get unified gene and transcript table\n",
    "gene_data = gene_data.loc[:, ['gene_id'] + samples ]\n",
    "transcript_data = transcript_data.loc[:,['gene_id', 'transcript_id'] + samples]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "f5976855",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>C00039</th>\n",
       "      <th>C00055</th>\n",
       "      <th>C00074</th>\n",
       "      <th>C00085</th>\n",
       "      <th>C00105</th>\n",
       "      <th>C00120</th>\n",
       "      <th>C00132</th>\n",
       "      <th>C00158</th>\n",
       "      <th>...</th>\n",
       "      <th>P01600</th>\n",
       "      <th>P01603</th>\n",
       "      <th>P01610</th>\n",
       "      <th>P01616</th>\n",
       "      <th>P01621</th>\n",
       "      <th>P01622</th>\n",
       "      <th>P01623</th>\n",
       "      <th>P01626</th>\n",
       "      <th>P01628</th>\n",
       "      <th>P01635</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000244682</td>\n",
       "      <td>ENST00000473530</td>\n",
       "      <td>2.276150</td>\n",
       "      <td>2.040930</td>\n",
       "      <td>1.564230</td>\n",
       "      <td>2.629880</td>\n",
       "      <td>3.224940</td>\n",
       "      <td>0.081621</td>\n",
       "      <td>3.577190</td>\n",
       "      <td>3.275500</td>\n",
       "      <td>...</td>\n",
       "      <td>1.130160</td>\n",
       "      <td>0.495594</td>\n",
       "      <td>1.458520</td>\n",
       "      <td>2.579210</td>\n",
       "      <td>0.110632</td>\n",
       "      <td>1.342680</td>\n",
       "      <td>2.268880</td>\n",
       "      <td>2.249030</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.658050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000244682</td>\n",
       "      <td>ENST00000467903</td>\n",
       "      <td>2.167410</td>\n",
       "      <td>2.788660</td>\n",
       "      <td>3.829120</td>\n",
       "      <td>3.177140</td>\n",
       "      <td>4.211690</td>\n",
       "      <td>0.176553</td>\n",
       "      <td>6.959880</td>\n",
       "      <td>3.485350</td>\n",
       "      <td>...</td>\n",
       "      <td>2.218410</td>\n",
       "      <td>1.574970</td>\n",
       "      <td>2.190090</td>\n",
       "      <td>2.435190</td>\n",
       "      <td>0.364871</td>\n",
       "      <td>2.658690</td>\n",
       "      <td>2.882860</td>\n",
       "      <td>4.777670</td>\n",
       "      <td>0.256739</td>\n",
       "      <td>2.298990</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000244682</td>\n",
       "      <td>ENST00000508651</td>\n",
       "      <td>1.950990</td>\n",
       "      <td>2.569730</td>\n",
       "      <td>3.989870</td>\n",
       "      <td>3.134920</td>\n",
       "      <td>5.018950</td>\n",
       "      <td>0.110165</td>\n",
       "      <td>4.027630</td>\n",
       "      <td>1.973540</td>\n",
       "      <td>...</td>\n",
       "      <td>3.231000</td>\n",
       "      <td>1.006310</td>\n",
       "      <td>0.281987</td>\n",
       "      <td>2.741400</td>\n",
       "      <td>0.052092</td>\n",
       "      <td>2.602130</td>\n",
       "      <td>3.153650</td>\n",
       "      <td>4.473280</td>\n",
       "      <td>0.185469</td>\n",
       "      <td>1.629250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000280938</td>\n",
       "      <td>ENST00000626647</td>\n",
       "      <td>0.049179</td>\n",
       "      <td>0.019637</td>\n",
       "      <td>0.017334</td>\n",
       "      <td>0.489663</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.105393</td>\n",
       "      <td>0.517216</td>\n",
       "      <td>0.029199</td>\n",
       "      <td>...</td>\n",
       "      <td>0.234445</td>\n",
       "      <td>0.072602</td>\n",
       "      <td>0.436067</td>\n",
       "      <td>0.057276</td>\n",
       "      <td>0.373421</td>\n",
       "      <td>0.103772</td>\n",
       "      <td>0.383861</td>\n",
       "      <td>0.494979</td>\n",
       "      <td>6.911700</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000282181</td>\n",
       "      <td>ENST00000633377</td>\n",
       "      <td>3.231750</td>\n",
       "      <td>1.134710</td>\n",
       "      <td>3.935840</td>\n",
       "      <td>0.548367</td>\n",
       "      <td>0.007815</td>\n",
       "      <td>0.319113</td>\n",
       "      <td>3.020600</td>\n",
       "      <td>1.936820</td>\n",
       "      <td>...</td>\n",
       "      <td>0.478180</td>\n",
       "      <td>0.111222</td>\n",
       "      <td>0.650404</td>\n",
       "      <td>0.288542</td>\n",
       "      <td>0.006630</td>\n",
       "      <td>1.255300</td>\n",
       "      <td>1.448720</td>\n",
       "      <td>2.497850</td>\n",
       "      <td>0.007706</td>\n",
       "      <td>1.090220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39310</th>\n",
       "      <td>ENSG00000197081</td>\n",
       "      <td>ENST00000569097</td>\n",
       "      <td>0.821758</td>\n",
       "      <td>0.377072</td>\n",
       "      <td>0.413925</td>\n",
       "      <td>0.907076</td>\n",
       "      <td>0.476571</td>\n",
       "      <td>0.098923</td>\n",
       "      <td>0.395161</td>\n",
       "      <td>0.379031</td>\n",
       "      <td>...</td>\n",
       "      <td>0.434826</td>\n",
       "      <td>0.292465</td>\n",
       "      <td>0.449118</td>\n",
       "      <td>0.594457</td>\n",
       "      <td>0.060498</td>\n",
       "      <td>0.269001</td>\n",
       "      <td>0.681752</td>\n",
       "      <td>0.957061</td>\n",
       "      <td>0.047140</td>\n",
       "      <td>0.631274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39311</th>\n",
       "      <td>ENSG00000225366</td>\n",
       "      <td>ENST00000602699</td>\n",
       "      <td>0.050290</td>\n",
       "      <td>0.014330</td>\n",
       "      <td>0.035171</td>\n",
       "      <td>0.041823</td>\n",
       "      <td>0.037060</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.065104</td>\n",
       "      <td>0.016308</td>\n",
       "      <td>...</td>\n",
       "      <td>0.051768</td>\n",
       "      <td>0.040899</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.036091</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.022102</td>\n",
       "      <td>0.044203</td>\n",
       "      <td>0.044029</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.051400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39312</th>\n",
       "      <td>ENSG00000182986</td>\n",
       "      <td>ENST00000594408</td>\n",
       "      <td>2.056060</td>\n",
       "      <td>1.068780</td>\n",
       "      <td>0.909087</td>\n",
       "      <td>1.211280</td>\n",
       "      <td>1.235250</td>\n",
       "      <td>0.022664</td>\n",
       "      <td>1.194140</td>\n",
       "      <td>1.571050</td>\n",
       "      <td>...</td>\n",
       "      <td>2.038440</td>\n",
       "      <td>0.777621</td>\n",
       "      <td>0.761086</td>\n",
       "      <td>1.157570</td>\n",
       "      <td>0.064273</td>\n",
       "      <td>0.759715</td>\n",
       "      <td>1.245030</td>\n",
       "      <td>1.833670</td>\n",
       "      <td>0.017389</td>\n",
       "      <td>1.783930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39313</th>\n",
       "      <td>ENSG00000205482</td>\n",
       "      <td>ENST00000510091</td>\n",
       "      <td>0.397195</td>\n",
       "      <td>0.139399</td>\n",
       "      <td>0.137387</td>\n",
       "      <td>0.238294</td>\n",
       "      <td>0.456423</td>\n",
       "      <td>0.055239</td>\n",
       "      <td>0.127514</td>\n",
       "      <td>0.330749</td>\n",
       "      <td>...</td>\n",
       "      <td>0.067772</td>\n",
       "      <td>0.278306</td>\n",
       "      <td>0.285876</td>\n",
       "      <td>0.041657</td>\n",
       "      <td>0.033892</td>\n",
       "      <td>0.187877</td>\n",
       "      <td>0.468567</td>\n",
       "      <td>1.001040</td>\n",
       "      <td>0.049527</td>\n",
       "      <td>0.233459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39314</th>\n",
       "      <td>ENSG00000184616</td>\n",
       "      <td>ENST00000618416</td>\n",
       "      <td>0.769121</td>\n",
       "      <td>0.759989</td>\n",
       "      <td>0.346327</td>\n",
       "      <td>0.559473</td>\n",
       "      <td>0.161998</td>\n",
       "      <td>0.323150</td>\n",
       "      <td>1.094700</td>\n",
       "      <td>0.756638</td>\n",
       "      <td>...</td>\n",
       "      <td>0.592340</td>\n",
       "      <td>0.360954</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.103718</td>\n",
       "      <td>0.098629</td>\n",
       "      <td>0.434144</td>\n",
       "      <td>0.721068</td>\n",
       "      <td>0.476099</td>\n",
       "      <td>0.116454</td>\n",
       "      <td>0.425519</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>39315 rows × 168 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene_id    transcript_id    C00039    C00055    C00074  \\\n",
       "0      ENSG00000244682  ENST00000473530  2.276150  2.040930  1.564230   \n",
       "1      ENSG00000244682  ENST00000467903  2.167410  2.788660  3.829120   \n",
       "2      ENSG00000244682  ENST00000508651  1.950990  2.569730  3.989870   \n",
       "3      ENSG00000280938  ENST00000626647  0.049179  0.019637  0.017334   \n",
       "4      ENSG00000282181  ENST00000633377  3.231750  1.134710  3.935840   \n",
       "...                ...              ...       ...       ...       ...   \n",
       "39310  ENSG00000197081  ENST00000569097  0.821758  0.377072  0.413925   \n",
       "39311  ENSG00000225366  ENST00000602699  0.050290  0.014330  0.035171   \n",
       "39312  ENSG00000182986  ENST00000594408  2.056060  1.068780  0.909087   \n",
       "39313  ENSG00000205482  ENST00000510091  0.397195  0.139399  0.137387   \n",
       "39314  ENSG00000184616  ENST00000618416  0.769121  0.759989  0.346327   \n",
       "\n",
       "         C00085    C00105    C00120    C00132    C00158  ...    P01600  \\\n",
       "0      2.629880  3.224940  0.081621  3.577190  3.275500  ...  1.130160   \n",
       "1      3.177140  4.211690  0.176553  6.959880  3.485350  ...  2.218410   \n",
       "2      3.134920  5.018950  0.110165  4.027630  1.973540  ...  3.231000   \n",
       "3      0.489663  0.000000  0.105393  0.517216  0.029199  ...  0.234445   \n",
       "4      0.548367  0.007815  0.319113  3.020600  1.936820  ...  0.478180   \n",
       "...         ...       ...       ...       ...       ...  ...       ...   \n",
       "39310  0.907076  0.476571  0.098923  0.395161  0.379031  ...  0.434826   \n",
       "39311  0.041823  0.037060  0.000000  0.065104  0.016308  ...  0.051768   \n",
       "39312  1.211280  1.235250  0.022664  1.194140  1.571050  ...  2.038440   \n",
       "39313  0.238294  0.456423  0.055239  0.127514  0.330749  ...  0.067772   \n",
       "39314  0.559473  0.161998  0.323150  1.094700  0.756638  ...  0.592340   \n",
       "\n",
       "         P01603    P01610    P01616    P01621    P01622    P01623    P01626  \\\n",
       "0      0.495594  1.458520  2.579210  0.110632  1.342680  2.268880  2.249030   \n",
       "1      1.574970  2.190090  2.435190  0.364871  2.658690  2.882860  4.777670   \n",
       "2      1.006310  0.281987  2.741400  0.052092  2.602130  3.153650  4.473280   \n",
       "3      0.072602  0.436067  0.057276  0.373421  0.103772  0.383861  0.494979   \n",
       "4      0.111222  0.650404  0.288542  0.006630  1.255300  1.448720  2.497850   \n",
       "...         ...       ...       ...       ...       ...       ...       ...   \n",
       "39310  0.292465  0.449118  0.594457  0.060498  0.269001  0.681752  0.957061   \n",
       "39311  0.040899  0.000000  0.036091  0.000000  0.022102  0.044203  0.044029   \n",
       "39312  0.777621  0.761086  1.157570  0.064273  0.759715  1.245030  1.833670   \n",
       "39313  0.278306  0.285876  0.041657  0.033892  0.187877  0.468567  1.001040   \n",
       "39314  0.360954  0.000000  0.103718  0.098629  0.434144  0.721068  0.476099   \n",
       "\n",
       "         P01628    P01635  \n",
       "0      0.000000  2.658050  \n",
       "1      0.256739  2.298990  \n",
       "2      0.185469  1.629250  \n",
       "3      6.911700  0.000000  \n",
       "4      0.007706  1.090220  \n",
       "...         ...       ...  \n",
       "39310  0.047140  0.631274  \n",
       "39311  0.000000  0.051400  \n",
       "39312  0.017389  1.783930  \n",
       "39313  0.049527  0.233459  \n",
       "39314  0.116454  0.425519  \n",
       "\n",
       "[39315 rows x 168 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcript_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "72d91038",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene stable ID version</th>\n",
       "      <th>Transcript stable ID</th>\n",
       "      <th>Transcript stable ID version</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Transcript name</th>\n",
       "      <th>Gene type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>ENSG00000210049.1</td>\n",
       "      <td>ENST00000387314</td>\n",
       "      <td>ENST00000387314.1</td>\n",
       "      <td>MT-TF</td>\n",
       "      <td>MT-TF-201</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>ENSG00000211459.2</td>\n",
       "      <td>ENST00000389680</td>\n",
       "      <td>ENST00000389680.2</td>\n",
       "      <td>MT-RNR1</td>\n",
       "      <td>MT-RNR1-201</td>\n",
       "      <td>Mt_rRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>ENSG00000210077.1</td>\n",
       "      <td>ENST00000387342</td>\n",
       "      <td>ENST00000387342.1</td>\n",
       "      <td>MT-TV</td>\n",
       "      <td>MT-TV-201</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>ENSG00000210082.2</td>\n",
       "      <td>ENST00000387347</td>\n",
       "      <td>ENST00000387347.2</td>\n",
       "      <td>MT-RNR2</td>\n",
       "      <td>MT-RNR2-201</td>\n",
       "      <td>Mt_rRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000209082</td>\n",
       "      <td>ENSG00000209082.1</td>\n",
       "      <td>ENST00000386347</td>\n",
       "      <td>ENST00000386347.1</td>\n",
       "      <td>MT-TL1</td>\n",
       "      <td>MT-TL1-201</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>278215</th>\n",
       "      <td>ENSG00000228437</td>\n",
       "      <td>ENSG00000228437.5</td>\n",
       "      <td>ENST00000441160</td>\n",
       "      <td>ENST00000441160.1</td>\n",
       "      <td>LINC02474</td>\n",
       "      <td>LINC02474-202</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>278216</th>\n",
       "      <td>ENSG00000229463</td>\n",
       "      <td>ENSG00000229463.2</td>\n",
       "      <td>ENST00000412098</td>\n",
       "      <td>ENST00000412098.2</td>\n",
       "      <td>LYST-AS1</td>\n",
       "      <td>LYST-AS1-201</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>278217</th>\n",
       "      <td>ENSG00000229291</td>\n",
       "      <td>ENSG00000229291.2</td>\n",
       "      <td>ENST00000438371</td>\n",
       "      <td>ENST00000438371.1</td>\n",
       "      <td>LINC02768</td>\n",
       "      <td>LINC02768-201</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>278218</th>\n",
       "      <td>ENSG00000229291</td>\n",
       "      <td>ENSG00000229291.2</td>\n",
       "      <td>ENST00000715678</td>\n",
       "      <td>ENST00000715678.1</td>\n",
       "      <td>LINC02768</td>\n",
       "      <td>LINC02768-203</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>278219</th>\n",
       "      <td>ENSG00000229291</td>\n",
       "      <td>ENSG00000229291.2</td>\n",
       "      <td>ENST00000443207</td>\n",
       "      <td>ENST00000443207.1</td>\n",
       "      <td>LINC02768</td>\n",
       "      <td>LINC02768-202</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>278220 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Gene stable ID Gene stable ID version Transcript stable ID  \\\n",
       "0       ENSG00000210049      ENSG00000210049.1      ENST00000387314   \n",
       "1       ENSG00000211459      ENSG00000211459.2      ENST00000389680   \n",
       "2       ENSG00000210077      ENSG00000210077.1      ENST00000387342   \n",
       "3       ENSG00000210082      ENSG00000210082.2      ENST00000387347   \n",
       "4       ENSG00000209082      ENSG00000209082.1      ENST00000386347   \n",
       "...                 ...                    ...                  ...   \n",
       "278215  ENSG00000228437      ENSG00000228437.5      ENST00000441160   \n",
       "278216  ENSG00000229463      ENSG00000229463.2      ENST00000412098   \n",
       "278217  ENSG00000229291      ENSG00000229291.2      ENST00000438371   \n",
       "278218  ENSG00000229291      ENSG00000229291.2      ENST00000715678   \n",
       "278219  ENSG00000229291      ENSG00000229291.2      ENST00000443207   \n",
       "\n",
       "       Transcript stable ID version  Gene name Transcript name Gene type  \n",
       "0                 ENST00000387314.1      MT-TF       MT-TF-201   Mt_tRNA  \n",
       "1                 ENST00000389680.2    MT-RNR1     MT-RNR1-201   Mt_rRNA  \n",
       "2                 ENST00000387342.1      MT-TV       MT-TV-201   Mt_tRNA  \n",
       "3                 ENST00000387347.2    MT-RNR2     MT-RNR2-201   Mt_rRNA  \n",
       "4                 ENST00000386347.1     MT-TL1      MT-TL1-201   Mt_tRNA  \n",
       "...                             ...        ...             ...       ...  \n",
       "278215            ENST00000441160.1  LINC02474   LINC02474-202    lncRNA  \n",
       "278216            ENST00000412098.2   LYST-AS1    LYST-AS1-201    lncRNA  \n",
       "278217            ENST00000438371.1  LINC02768   LINC02768-201    lncRNA  \n",
       "278218            ENST00000715678.1  LINC02768   LINC02768-203    lncRNA  \n",
       "278219            ENST00000443207.1  LINC02768   LINC02768-202    lncRNA  \n",
       "\n",
       "[278220 rows x 7 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biomart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "32fdf04b",
   "metadata": {},
   "outputs": [],
   "source": [
    "tf_iso_counts, _ = separate_tf_genes(data=transcript_data, tf_list=tf_list, data_column='transcript_id', biomart_column='Transcript stable ID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "cf58253a",
   "metadata": {},
   "outputs": [],
   "source": [
    "tf_gene_counts, target_gene_counts = separate_tf_genes(data=gene_data, tf_list=tf_list, data_column='gene_id', biomart_column='Gene stable ID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "56eea2b8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Preparing data for inference: Separation into canonical and as-aware gene expression data\n"
     ]
    }
   ],
   "source": [
    "data_canonical, data_asware, target_gene_list = prepare_for_inference(tf_iso_counts, tf_gene_counts, target_gene_counts, transcript_column='transcript_id', gene_column='gene_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "8a4e35eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get isoform categories\n",
    "isoform_categories = isoform_categorization(tf_iso_counts, tf_gene_counts, threshold_dominance=20)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "7933898d",
   "metadata": {},
   "outputs": [],
   "source": [
    "from inference import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "da84dd89",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting inference ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data_nfs/og86asub/alternet-project/alternet/.pixi/envs/default/lib/python3.11/site-packages/distributed/node.py:187: UserWarning: Port 8787 is already in use.\n",
      "Perhaps you already have a cluster running?\n",
      "Hosting the HTTP server on port 33487 instead\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "\n",
      "Computing network\n"
     ]
    },
    {
     "ename": "PermissionError",
     "evalue": "[Errno 13] Permission denied: '/data_nfs/mi34qaba/SpliceAwareGRN/results/NF/grn/NF_as-aware.network_1.tsv'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mPermissionError\u001b[39m                           Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[43]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m as_aware_grn, canonical_grn = \u001b[43minference\u001b[49m\u001b[43m(\u001b[49m\u001b[43mconfig\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[32;43m1\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtf_list\u001b[49m\u001b[43m=\u001b[49m\u001b[43mtf_list\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdata_canonical\u001b[49m\u001b[43m=\u001b[49m\u001b[43mdata_canonical\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdata_asware\u001b[49m\u001b[43m=\u001b[49m\u001b[43mdata_asware\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtarget_gene_list\u001b[49m\u001b[43m=\u001b[49m\u001b[43mtarget_gene_list\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maggregate\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/data_nfs/og86asub/alternet-project/alternet/src/alternet/inference.py:99\u001b[39m, in \u001b[36minference\u001b[39m\u001b[34m(config, nruns, data_canonical, data_asware, target_gene_list, tf_list, aggregate)\u001b[39m\n\u001b[32m     97\u001b[39m     \u001b[38;5;28mprint\u001b[39m()\n\u001b[32m     98\u001b[39m     file = op.join(results_dir_grn, as_aware_prefix + \u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mi\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m.tsv\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m---> \u001b[39m\u001b[32m99\u001b[39m     grn1 = \u001b[43mcompute_and_save_network\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdata_asaware_t\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    100\u001b[39m \u001b[43m                                \u001b[49m\u001b[43mtarget_gene_list\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    101\u001b[39m \u001b[43m                                \u001b[49m\u001b[43mtfs_transcripts\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    102\u001b[39m \u001b[43m                                \u001b[49m\u001b[43mclient\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    103\u001b[39m \u001b[43m                                \u001b[49m\u001b[43mfile\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    104\u001b[39m \u001b[43m                                \u001b[49m\u001b[43muse_tf\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[32m    107\u001b[39m \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(\u001b[32m1\u001b[39m, nruns+\u001b[32m1\u001b[39m):\n\u001b[32m    108\u001b[39m     \u001b[38;5;28mprint\u001b[39m(i)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/data_nfs/og86asub/alternet-project/alternet/src/alternet/inference.py:43\u001b[39m, in \u001b[36mcompute_and_save_network\u001b[39m\u001b[34m(gene_data, target_names, tf_list, client, file, use_tf)\u001b[39m\n\u001b[32m     41\u001b[39m \u001b[38;5;66;03m# write the GRN to file\u001b[39;00m\n\u001b[32m     42\u001b[39m network.columns = [\u001b[33m'\u001b[39m\u001b[33msource\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mtarget\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mimportance\u001b[39m\u001b[33m'\u001b[39m]\n\u001b[32m---> \u001b[39m\u001b[32m43\u001b[39m \u001b[43mnetwork\u001b[49m\u001b[43m.\u001b[49m\u001b[43mto_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfile\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msep\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[38;5;130;43;01m\\t\u001b[39;49;00m\u001b[33;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindex\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mheader\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[32m     45\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m network\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/data_nfs/og86asub/alternet-project/alternet/.pixi/envs/default/lib/python3.11/site-packages/pandas/util/_decorators.py:333\u001b[39m, in \u001b[36mdeprecate_nonkeyword_arguments.<locals>.decorate.<locals>.wrapper\u001b[39m\u001b[34m(*args, **kwargs)\u001b[39m\n\u001b[32m    327\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(args) > num_allow_args:\n\u001b[32m    328\u001b[39m     warnings.warn(\n\u001b[32m    329\u001b[39m         msg.format(arguments=_format_argument_list(allow_args)),\n\u001b[32m    330\u001b[39m         \u001b[38;5;167;01mFutureWarning\u001b[39;00m,\n\u001b[32m    331\u001b[39m         stacklevel=find_stack_level(),\n\u001b[32m    332\u001b[39m     )\n\u001b[32m--> \u001b[39m\u001b[32m333\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[43m*\u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/data_nfs/og86asub/alternet-project/alternet/.pixi/envs/default/lib/python3.11/site-packages/pandas/core/generic.py:3986\u001b[39m, in \u001b[36mNDFrame.to_csv\u001b[39m\u001b[34m(self, path_or_buf, sep, na_rep, float_format, columns, header, index, index_label, mode, encoding, compression, quoting, quotechar, lineterminator, chunksize, date_format, doublequote, escapechar, decimal, errors, storage_options)\u001b[39m\n\u001b[32m   3975\u001b[39m df = \u001b[38;5;28mself\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(\u001b[38;5;28mself\u001b[39m, ABCDataFrame) \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28mself\u001b[39m.to_frame()\n\u001b[32m   3977\u001b[39m formatter = DataFrameFormatter(\n\u001b[32m   3978\u001b[39m     frame=df,\n\u001b[32m   3979\u001b[39m     header=header,\n\u001b[32m   (...)\u001b[39m\u001b[32m   3983\u001b[39m     decimal=decimal,\n\u001b[32m   3984\u001b[39m )\n\u001b[32m-> \u001b[39m\u001b[32m3986\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mDataFrameRenderer\u001b[49m\u001b[43m(\u001b[49m\u001b[43mformatter\u001b[49m\u001b[43m)\u001b[49m\u001b[43m.\u001b[49m\u001b[43mto_csv\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m   3987\u001b[39m \u001b[43m    \u001b[49m\u001b[43mpath_or_buf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3988\u001b[39m \u001b[43m    \u001b[49m\u001b[43mlineterminator\u001b[49m\u001b[43m=\u001b[49m\u001b[43mlineterminator\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3989\u001b[39m \u001b[43m    \u001b[49m\u001b[43msep\u001b[49m\u001b[43m=\u001b[49m\u001b[43msep\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3990\u001b[39m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[43m=\u001b[49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3991\u001b[39m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[43m=\u001b[49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3992\u001b[39m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[43m=\u001b[49m\u001b[43mcompression\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3993\u001b[39m \u001b[43m    \u001b[49m\u001b[43mquoting\u001b[49m\u001b[43m=\u001b[49m\u001b[43mquoting\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3994\u001b[39m \u001b[43m    \u001b[49m\u001b[43mcolumns\u001b[49m\u001b[43m=\u001b[49m\u001b[43mcolumns\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3995\u001b[39m \u001b[43m    \u001b[49m\u001b[43mindex_label\u001b[49m\u001b[43m=\u001b[49m\u001b[43mindex_label\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3996\u001b[39m \u001b[43m    \u001b[49m\u001b[43mmode\u001b[49m\u001b[43m=\u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3997\u001b[39m \u001b[43m    \u001b[49m\u001b[43mchunksize\u001b[49m\u001b[43m=\u001b[49m\u001b[43mchunksize\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3998\u001b[39m \u001b[43m    \u001b[49m\u001b[43mquotechar\u001b[49m\u001b[43m=\u001b[49m\u001b[43mquotechar\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3999\u001b[39m \u001b[43m    \u001b[49m\u001b[43mdate_format\u001b[49m\u001b[43m=\u001b[49m\u001b[43mdate_format\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   4000\u001b[39m \u001b[43m    \u001b[49m\u001b[43mdoublequote\u001b[49m\u001b[43m=\u001b[49m\u001b[43mdoublequote\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   4001\u001b[39m \u001b[43m    \u001b[49m\u001b[43mescapechar\u001b[49m\u001b[43m=\u001b[49m\u001b[43mescapechar\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   4002\u001b[39m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[43m=\u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   4003\u001b[39m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/data_nfs/og86asub/alternet-project/alternet/.pixi/envs/default/lib/python3.11/site-packages/pandas/io/formats/format.py:1014\u001b[39m, in \u001b[36mDataFrameRenderer.to_csv\u001b[39m\u001b[34m(self, path_or_buf, encoding, sep, columns, index_label, mode, compression, quoting, quotechar, lineterminator, chunksize, date_format, doublequote, escapechar, errors, storage_options)\u001b[39m\n\u001b[32m    993\u001b[39m     created_buffer = \u001b[38;5;28;01mFalse\u001b[39;00m\n\u001b[32m    995\u001b[39m csv_formatter = CSVFormatter(\n\u001b[32m    996\u001b[39m     path_or_buf=path_or_buf,\n\u001b[32m    997\u001b[39m     lineterminator=lineterminator,\n\u001b[32m   (...)\u001b[39m\u001b[32m   1012\u001b[39m     formatter=\u001b[38;5;28mself\u001b[39m.fmt,\n\u001b[32m   1013\u001b[39m )\n\u001b[32m-> \u001b[39m\u001b[32m1014\u001b[39m \u001b[43mcsv_formatter\u001b[49m\u001b[43m.\u001b[49m\u001b[43msave\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   1016\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m created_buffer:\n\u001b[32m   1017\u001b[39m     \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(path_or_buf, StringIO)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/data_nfs/og86asub/alternet-project/alternet/.pixi/envs/default/lib/python3.11/site-packages/pandas/io/formats/csvs.py:251\u001b[39m, in \u001b[36mCSVFormatter.save\u001b[39m\u001b[34m(self)\u001b[39m\n\u001b[32m    247\u001b[39m \u001b[38;5;250m\u001b[39m\u001b[33;03m\"\"\"\u001b[39;00m\n\u001b[32m    248\u001b[39m \u001b[33;03mCreate the writer & save.\u001b[39;00m\n\u001b[32m    249\u001b[39m \u001b[33;03m\"\"\"\u001b[39;00m\n\u001b[32m    250\u001b[39m \u001b[38;5;66;03m# apply compression and byte/text conversion\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m251\u001b[39m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[43mget_handle\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m    252\u001b[39m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    253\u001b[39m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    254\u001b[39m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    255\u001b[39m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    256\u001b[39m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mcompression\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    257\u001b[39m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    258\u001b[39m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mas\u001b[39;00m handles:\n\u001b[32m    259\u001b[39m     \u001b[38;5;66;03m# Note: self.encoding is irrelevant here\u001b[39;00m\n\u001b[32m    260\u001b[39m     \u001b[38;5;28mself\u001b[39m.writer = csvlib.writer(\n\u001b[32m    261\u001b[39m         handles.handle,\n\u001b[32m    262\u001b[39m         lineterminator=\u001b[38;5;28mself\u001b[39m.lineterminator,\n\u001b[32m   (...)\u001b[39m\u001b[32m    267\u001b[39m         quotechar=\u001b[38;5;28mself\u001b[39m.quotechar,\n\u001b[32m    268\u001b[39m     )\n\u001b[32m    270\u001b[39m     \u001b[38;5;28mself\u001b[39m._save()\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/data_nfs/og86asub/alternet-project/alternet/.pixi/envs/default/lib/python3.11/site-packages/pandas/io/common.py:873\u001b[39m, in \u001b[36mget_handle\u001b[39m\u001b[34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[39m\n\u001b[32m    868\u001b[39m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(handle, \u001b[38;5;28mstr\u001b[39m):\n\u001b[32m    869\u001b[39m     \u001b[38;5;66;03m# Check whether the filename is to be opened in binary mode.\u001b[39;00m\n\u001b[32m    870\u001b[39m     \u001b[38;5;66;03m# Binary mode does not support 'encoding' and 'newline'.\u001b[39;00m\n\u001b[32m    871\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m ioargs.encoding \u001b[38;5;129;01mand\u001b[39;00m \u001b[33m\"\u001b[39m\u001b[33mb\u001b[39m\u001b[33m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m ioargs.mode:\n\u001b[32m    872\u001b[39m         \u001b[38;5;66;03m# Encoding\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m873\u001b[39m         handle = \u001b[38;5;28mopen\u001b[39m(\n\u001b[32m    874\u001b[39m             handle,\n\u001b[32m    875\u001b[39m             ioargs.mode,\n\u001b[32m    876\u001b[39m             encoding=ioargs.encoding,\n\u001b[32m    877\u001b[39m             errors=errors,\n\u001b[32m    878\u001b[39m             newline=\u001b[33m\"\u001b[39m\u001b[33m\"\u001b[39m,\n\u001b[32m    879\u001b[39m         )\n\u001b[32m    880\u001b[39m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m    881\u001b[39m         \u001b[38;5;66;03m# Binary mode\u001b[39;00m\n\u001b[32m    882\u001b[39m         handle = \u001b[38;5;28mopen\u001b[39m(handle, ioargs.mode)\n",
      "\u001b[31mPermissionError\u001b[39m: [Errno 13] Permission denied: '/data_nfs/mi34qaba/SpliceAwareGRN/results/NF/grn/NF_as-aware.network_1.tsv'"
     ]
    }
   ],
   "source": [
    "as_aware_grn, canonical_grn = inference(config, 1, tf_list=tf_list, data_canonical=data_canonical, data_asware=data_asware, target_gene_list=target_gene_list, aggregate=True)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pixi_alternet",
   "language": "python",
   "name": "pixi_alternet"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
